{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11809fe2",
   "metadata": {},
   "source": [
    "# Calculating DCJ distance between two signed, circular and/or linear genomes with no duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ea4c69",
   "metadata": {},
   "source": [
    "Credit:\n",
    "- data structures from https://github.com/mlliou112/py-dcj \n",
    "- ChatGPT for converting to Julia "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "221c64a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Base.Iterators\n",
    "using Parameters\n",
    "using DataStructures\n",
    "\n",
    "abstract type AbstractGene end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "91f5f99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# telomere\n",
    "\n",
    "struct Telomere <: AbstractGene\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Telomere, b)\n",
    "    return typeof(a) == typeof(b)\n",
    "end\n",
    "\n",
    "function Base.show(t::Telomere)\n",
    "    print(\".\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b81dd52d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "return_ge_str (generic function with 1 method)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene\n",
    "\n",
    "@with_kw struct Gene <: AbstractGene\n",
    "    id::Int\n",
    "    dna::String  # lowercase char \n",
    "    dup::Int\n",
    "    reverse::Bool=false\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Gene, b::Gene)\n",
    "    return typeof(a) == typeof(b) && a.id == b.id\n",
    "end\n",
    "\n",
    "function Base.show(gene::Gene)\n",
    "    g = gene.dna\n",
    "    if gene.reverse\n",
    "        g = uppercase(g)\n",
    "    end \n",
    "\n",
    "    print(g*string(gene.dup))\n",
    "end\n",
    "\n",
    "function return_ge_str(gene::Gene)\n",
    "    g = gene.dna\n",
    "    if gene.reverse\n",
    "        g = uppercase(g)\n",
    "    end \n",
    "\n",
    "    return (g*string(gene.dup))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b1a98bd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_genes (generic function with 1 method)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# str is ::SubString{String} || ::String\n",
    "function find_genes(str)\n",
    "    genes = String[]\n",
    "    curr_gene = \"\"\n",
    "\n",
    "    seen_ints = false \n",
    "\n",
    "    for g in str\n",
    "        if g == '.'  # telomere \n",
    "            if !isempty(curr_gene) \n",
    "                push!(genes, curr_gene) \n",
    "            end\n",
    "            push!(genes, string(g))\n",
    "            curr_gene = \"\"\n",
    "        else   # non-telomere \n",
    "            if isdigit(g) && !seen_ints\n",
    "                seen_ints = true \n",
    "            elseif seen_ints && isletter(g)\n",
    "                if !isempty(curr_gene) \n",
    "                    push!(genes, curr_gene) \n",
    "                    seen_ints = false\n",
    "                end\n",
    "                curr_gene = \"\"\n",
    "            end \n",
    "            curr_gene *= g\n",
    "        end \n",
    "    end\n",
    "\n",
    "    # add last accumulated string if not empty\n",
    "    if !isempty(curr_gene) push!(genes, curr_gene) end \n",
    "\n",
    "    return genes\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f20e9ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chromosome\n",
    "\n",
    "mutable struct Chromosome\n",
    "    genes::Vector{AbstractGene}\n",
    "end\n",
    "\n",
    "function Base.show(chrom::Chromosome)\n",
    "    for gene in chrom.genes \n",
    "        show(gene)\n",
    "    end \n",
    "end \n",
    "\n",
    "function Chromosome(genes_str::SubString{String}, id_counter::Ref{Int64}, id_to_str::Dict{Int, String}, str_to_id::Dict{String, Int64}, target::Bool)\n",
    "    content = AbstractGene[]\n",
    "\n",
    "    genes = find_genes(genes_str)\n",
    "    \n",
    "    for (i, gene) in pairs(genes) \n",
    "        telomere = gene == \".\"   \n",
    "        if telomere && (i == 1 || i == length(genes))\n",
    "            push!(content, Telomere())\n",
    "        elseif telomere \n",
    "            throw(ArgumentError(\"Linear chromosome must start and end with telomeres.\"))\n",
    "        else  # gene \n",
    "            gene_letters = filter(isletter, gene)\n",
    "            rev = gene_letters == uppercase(gene_letters)\n",
    "            dna = lowercase(gene_letters)\n",
    "            dup = parse(Int, (filter(isdigit, gene)))\n",
    "            gene = dna*string(dup)\n",
    "            id = id_counter[]\n",
    "\n",
    "            if target  \n",
    "                id_to_str[id] = gene\n",
    "                if gene in keys(str_to_id)  \n",
    "                    throw(ArgumentError(\"Duplicate gene.\"))\n",
    "                else \n",
    "                    str_to_id[gene] = id\n",
    "                end \n",
    "                id_counter[] += 1\n",
    "\n",
    "            else  # source str \n",
    "                id = str_to_id[gene]\n",
    "            end\n",
    "            \n",
    "            push!(content, Gene(id, dna, dup, rev))\n",
    "        end \n",
    "    end    \n",
    "    return Chromosome(content)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0db29f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genome\n",
    "\n",
    "mutable struct Genome\n",
    "    data::Vector{Chromosome}\n",
    "    id_to_str::Dict{Int, String}\n",
    "    str_to_id::Dict{String, Int}\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Genome, b::Genome)\n",
    "    return a.data == a.data && a.id_to_str == b.id_to_str && a.str_to_id == b.str_to_id\n",
    "end\n",
    "\n",
    "function Base.show(genome::Genome)\n",
    "    for chrom in genome.data\n",
    "        show(chrom)\n",
    "    end \n",
    "end \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ef463477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_to_string (generic function with 1 method)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# str -> genome \n",
    "\n",
    "function string_to_genome(s, id_counter, id_to_str, str_to_id, target)\n",
    "    chrom_list_str = split(s, \",\")\n",
    "\n",
    "    # convert chrom_list_str (list of strings) to chrom_list\n",
    "    chrom_list = Chromosome[]\n",
    "    \n",
    "    for c_str in chrom_list_str\n",
    "        c = Chromosome(c_str, id_counter, id_to_str, str_to_id, target)\n",
    "        push!(chrom_list, c)\n",
    "    end \n",
    "    \n",
    "    return Genome(chrom_list, id_to_str, str_to_id)\n",
    "end \n",
    "\n",
    "\n",
    "function genome_to_string(genome::Genome)\n",
    "    genome_str = \"\"\n",
    "\n",
    "    chroms = genome.data \n",
    "\n",
    "    for c in chroms \n",
    "        genome_str *= c.genes[1]\n",
    "        for g in c.genes[2:end ]\n",
    "            genome_str *= \"'\"*g\n",
    "        end \n",
    "        genome_str *= \",\"\n",
    "    end \n",
    "\n",
    "    return genome_str \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9b91ea5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene end (used in adj)\n",
    "\n",
    "@with_kw struct GeneEnd\n",
    "    gene::AbstractGene\n",
    "    head::Bool=true\n",
    "end\n",
    "\n",
    "function GeneEnd(T::Telomere)\n",
    "    return GeneEnd(Telomere(), false)  # head value doesn't matter   \n",
    "end \n",
    "\n",
    "function Base.:(==)(a::GeneEnd, b::GeneEnd)\n",
    "    if a.gene == Telomere() && b.gene == Telomere() \n",
    "        return true \n",
    "    elseif a.gene == Telomere() || b.gene == Telomere() \n",
    "        return false \n",
    "    else \n",
    "        return a.gene.id == b.gene.id && a.head == b.head\n",
    "    end \n",
    "end\n",
    "\n",
    "function Base.show(ge::GeneEnd)\n",
    "    if ge.gene == Telomere() \n",
    "        return \".\"\n",
    "    end \n",
    "    \n",
    "    ge_info = ge.gene.dna * string(ge.gene.dup)\n",
    "    (ge.head == true) ? ge_info *= \":h\" : ge_info *= \":t\" \n",
    "\n",
    "    return ge_info\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4ef3f032",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjacency\n",
    "mutable struct Adjacency \n",
    "    left::GeneEnd\n",
    "    right::GeneEnd\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Adjacency, b::Adjacency)\n",
    "     # order of the gene ends doesn't matter\n",
    "    if a.left == b.left && a.right == b.right || a.left == b.right && a.right == b.left  \n",
    "        return true \n",
    "    end \n",
    "    return false \n",
    "end \n",
    "\n",
    "function Base.show(adj::Adjacency)\n",
    "    left = show(adj.left)\n",
    "    right = show(adj.right)\n",
    "\n",
    "    print(\"(\", left, \",\", right, \")\")\n",
    "end \n",
    "\n",
    "\n",
    "# logging \n",
    "\n",
    "function Base.show(adj::Adjacency, blue::Bool)\n",
    "    left = show(adj.left)\n",
    "    right = show(adj.right)\n",
    "    \n",
    "    s = \"(\" * left * \",\" * right * \")\"\n",
    "    \n",
    "    if blue \n",
    "        printstyled(s; color = :blue)\n",
    "    else\n",
    "        printstyled(s; color = :green) \n",
    "    end \n",
    "\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cb7a36eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# showing/printing list of adjs\n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency})\n",
    "    for adj in adj_list \n",
    "        show(adj) \n",
    "    end \n",
    "end \n",
    "\n",
    "\n",
    "# logging\n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency}, changed_adj1::Adjacency, changed_adj2::Adjacency)\n",
    "    for adj in adj_list \n",
    "        if adj == changed_adj1 || adj == changed_adj2\n",
    "            show(adj, true)\n",
    "        else \n",
    "            show(adj) \n",
    "        end \n",
    "    end \n",
    "end \n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency}, changed_adj::Adjacency) \n",
    "    for adj in adj_list \n",
    "        if adj == changed_adj\n",
    "            show(adj, true)\n",
    "        else \n",
    "            show(adj) \n",
    "        end \n",
    "    end \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f29ed684",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_to_adj_list (generic function with 1 method)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function genome_to_adj_list_helper(linear::Bool, chrom::Chromosome, genes::Vector{AbstractGene}, adj_list::Vector{Adjacency})\n",
    "    # first gene end \n",
    "    if linear \n",
    "        left = GeneEnd(Telomere())\n",
    "        end_idx = length(genes) - 1\n",
    "    else \n",
    "        left = GeneEnd(genes[1], !genes[1].reverse)\n",
    "        end_idx = length(genes)\n",
    "    end  \n",
    "\n",
    "    for i in 2:end_idx\n",
    "        gene = genes[i]\n",
    "        reversed = gene.reverse  \n",
    "        \n",
    "        right = GeneEnd(gene, reversed)  # if reversed, right_end_adj is head (& left_end_adj is tail) \n",
    "        adj = Adjacency(left, right)\n",
    "        push!(adj_list, adj)\n",
    "\n",
    "        left = GeneEnd(gene, !reversed)\n",
    "    end \n",
    "\n",
    "    # last gene end \n",
    "    if linear \n",
    "        right = GeneEnd(Telomere())\n",
    "    else \n",
    "        right = GeneEnd(genes[1], genes[1].reverse)\n",
    "    end  \n",
    "\n",
    "    adj = Adjacency(left, right)\n",
    "    push!(adj_list, adj)\n",
    "end \n",
    "\n",
    "function genome_to_adj_list(genome::Genome)\n",
    "    adj_list = Vector{Adjacency}()\n",
    "\n",
    "    for chrom in genome.data\n",
    "        genes = chrom.genes\n",
    "        if genes[1] == Telomere()  # linear \n",
    "            genome_to_adj_list_helper(true, chrom, genes, adj_list)\n",
    "        else  # circular \n",
    "            genome_to_adj_list_helper(false, chrom, genes, adj_list)\n",
    "        end\n",
    "    end \n",
    "    \n",
    "    return adj_list\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f9d96992",
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: `GeneEnd` not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: `GeneEnd` not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ In[3]:1"
     ]
    }
   ],
   "source": [
    "function common_gene_helper(common_ge::GeneEnd, adj::Adjacency)    \n",
    "    common_gene = common_ge.gene\n",
    "    if common_gene == adj.left.gene \n",
    "        if adj.left.head == true \n",
    "            return uppercase(common_gene.dna*string(common_gene.dup))\n",
    "        else \n",
    "            return lowercase(common_gene.dna*string(common_gene.dup))\n",
    "        end \n",
    "    else \n",
    "        if adj.right.head == true \n",
    "            return uppercase(common_gene.dna*string(common_gene.dup))\n",
    "        else \n",
    "            return lowercase(common_gene.dna*string(common_gene.dup))\n",
    "        end \n",
    "    end\n",
    "end \n",
    "\n",
    "function common_gene(adj1::Adjacency, adj2::Adjacency)\n",
    "    if adj1.left.gene == Telomere() \n",
    "        common_ge = adj1.right\n",
    "    elseif adj1.right.gene == Telomere() \n",
    "        common_ge = adj1.left \n",
    "    elseif adj1.left.gene == adj2.left.gene || adj1.left.gene == adj2.right.gene\n",
    "        common_ge = adj1.left\n",
    "    else \n",
    "        common_ge = adj1.right\n",
    "    end \n",
    "    letter  = common_gene_helper(common_ge, adj1) \n",
    "    \n",
    "    return letter, common_ge\n",
    "end \n",
    "\n",
    "function adjlist_to_genomestr(adj_list::Vector{Adjacency})\n",
    "    genome_str = \"\"\n",
    "\n",
    "    chrom = \"\"\n",
    "    circ = false \n",
    "    first_gene_circ = nothing \n",
    "    for i in range(1,length(adj_list))\n",
    "        # single adj contains one chrom (empty w telomeres or one gene) \n",
    "        if adj_list[i].left.gene == adj_list[i].right.gene \n",
    "            if adj_list[i].left.gene  == Telomere()\n",
    "                chrom =  \"..,\"\n",
    "                genome_str *= chrom \n",
    "                chrom = \"\"\n",
    "                continue\n",
    "            else \n",
    "                if adj_list[i].left.head == true \n",
    "                    gene = adj_list[i].left.gene\n",
    "                    uppercase(gene.dna*string(gene.dup))\n",
    "                else \n",
    "                    gene = adj_list[i].left.gene\n",
    "                    lowercase(gene.dna*string(gene.dup))\n",
    "                end \n",
    "                genome_str *= chrom * \",\"\n",
    "                chrom = \"\"\n",
    "                continue \n",
    "            end \n",
    "        end \n",
    "\n",
    "        # reached end \n",
    "        if i == length(adj_list) \n",
    "            if  (adj_list[i].left == GeneEnd(Telomere()) || adj_list[i].right == GeneEnd(Telomere()))\n",
    "                chrom *= \".\"\n",
    "                genome_str *= chrom \n",
    "                break\n",
    "            else # circular \n",
    "                # adj_list[i].left == first_gene_circ || adj_list[i].right == first_gene_circ\n",
    "                chrom *= common_gene_helper(first_gene_circ, adj_list[i]) \n",
    "                genome_str *= chrom \n",
    "                break\n",
    "            end \n",
    "        end \n",
    "\n",
    "        cg_char, cg_ge = common_gene(adj_list[i], adj_list[i+1]) \n",
    "        \n",
    "        if isempty(chrom) \n",
    "            if adj_list[i].left.gene != Telomere() &&  adj_list[i].right.gene != Telomere() \n",
    "                # starting new circ chrom \n",
    "                first_gene_circ = other_adjacency_end(cg_ge, adj_list[i])\n",
    "                circ = true \n",
    "            else # new linear chrom \n",
    "                chrom *= \".\"\n",
    "            end \n",
    "            chrom *= cg_char\n",
    "        else # adding to existing chrom \n",
    "            if circ \n",
    "                chrom *= cg_char\n",
    "                if other_adjacency_end(cg_ge, adj_list[i]) == first_gene_circ \n",
    "                    # end the circ chrom \n",
    "                    chrom *= first_gene_circ * \",\"\n",
    "                    genome_str *= chrom \n",
    "                    chrom = \"\"\n",
    "                end \n",
    "            else \n",
    "                if other_adjacency_end(cg_ge, adj_list[i]) == GeneEnd(Telomere()) \n",
    "                    chrom *= \".\" * \",\"\n",
    "                    genome_str *= chrom \n",
    "                    chrom = \"\"\n",
    "                else \n",
    "                    chrom *= cg_char\n",
    "                end \n",
    "            end \n",
    "        end \n",
    "    end \n",
    "    \n",
    "    return genome_str[1:end-1]\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b419e7",
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: `Genome` not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: `Genome` not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ In[1]:7"
     ]
    }
   ],
   "source": [
    "# adjacency graph \n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    num_dcj_ops::Int\n",
    "end\n",
    "\n",
    "function AdjacencyGraph(src::Genome, target::Genome, mode::Int)\n",
    "    # process genomes (not explicitly creating the adj graph)    \n",
    "    src_adj_list = genome_to_adj_list(src)        \n",
    "    target_adj_list  = genome_to_adj_list(target) \n",
    "    \n",
    "    if mode >= 1\n",
    "        show(src_adj_list)\n",
    "        print(\" --> \")\n",
    "        show(target_adj_list)\n",
    "        println()\n",
    "    end \n",
    "    \n",
    "    src_geneid_to_location = process_adj_list(src_adj_list)   \n",
    "    target_geneid_to_location = process_adj_list(target_adj_list)\n",
    "    \n",
    "    if mode >= 2\n",
    "        println()\n",
    "        print(\"SRC GENE & ID --> LOCATION DICT\\n\")\n",
    "        show(src_geneid_to_location, src.id_to_str)  \n",
    "        print(\"\\n\", \"TARGET GENE & ID --> LOCATION DICT\\n\")\n",
    "        show(target_geneid_to_location, src.id_to_str)\n",
    "        print(\"\\n*************\\n\\n\")\n",
    "    end \n",
    "\n",
    "    if src.id_to_str != target.id_to_str\n",
    "        throw(ArgumentError(\"Source and target genomes should have the same ID-->str dict\"))\n",
    "    end \n",
    "\n",
    "    num_dcj_ops = find_dcj_dist_ops(src_adj_list, target_adj_list, src_geneid_to_location, target_geneid_to_location, mode, target.id_to_str) \n",
    "    \n",
    "    if mode >= 1\n",
    "        print(\"\\n#ops = \", num_dcj_ops)\n",
    "        print(\"\\n----------------------------------------\\n\")\n",
    "    end \n",
    "\n",
    "    return AdjacencyGraph(num_dcj_ops)\n",
    "end "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
