{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11809fe2",
   "metadata": {},
   "source": [
    "# Calculating DCJ distance between two signed, circular and/or linear genomes with no duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ea4c69",
   "metadata": {},
   "source": [
    "Credit:\n",
    "- https://github.com/mlliou112/py-dcj (python version)\n",
    "- ChatGPT for converting to Julia "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "221c64a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Base.Iterators\n",
    "using Parameters\n",
    "using DataStructures\n",
    "\n",
    "abstract type AbstractGene end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "91f5f99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# telomere\n",
    "\n",
    "struct Telomere <: AbstractGene\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Telomere, b)\n",
    "    return typeof(a) == typeof(b)\n",
    "end\n",
    "\n",
    "function Base.show(t::Telomere)\n",
    "    return \".\"\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b81dd52d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene\n",
    "\n",
    "@with_kw struct Gene <: AbstractGene\n",
    "    id::Int\n",
    "    dna::Char  # lowercase \n",
    "    reverse::Bool=false\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Gene, b::Gene)\n",
    "    return typeof(a) == typeof(b) && a.id == b.id\n",
    "end\n",
    "\n",
    "function Base.show(m::Gene)\n",
    "    if m.reverse\n",
    "        print(uppercase(m.dna)) # * string(m.id))\n",
    "    else\n",
    "        print(m.dna) # string(m.id))\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7f20e9ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chromosome\n",
    "\n",
    "mutable struct Chromosome\n",
    "    genes::Vector{AbstractGene}\n",
    "end\n",
    "\n",
    "function Chromosome(genes::String, id_counter::Ref{Int64}, id_to_char::Dict{Int, Char}, char_to_id::Dict{Char, Int64}, target::Bool)\n",
    "    content = AbstractGene[]\n",
    "\n",
    "    for (i, gene) in pairs(genes)\n",
    "        dna = only(lowercase(string(gene)))\n",
    "        rev = isuppercase(gene)\n",
    "        telomere = gene == '.'    \n",
    "\n",
    "        if telomere \n",
    "            push!(content, Telomere())\n",
    "        else  # gene \n",
    "            id = id_counter[]\n",
    "\n",
    "            if target  \n",
    "                id_to_char[id] = gene\n",
    "                if gene in keys(char_to_id)  # TODO dups; char --> array of IDs  \n",
    "                    # push!(char_to_id[dna], id)\n",
    "                else \n",
    "                    char_to_id[dna] = id\n",
    "                end \n",
    "                id_counter[] += 1\n",
    "\n",
    "            else  # source str \n",
    "                ## assuming no duplicates\n",
    "                id = char_to_id[dna]  # TODO: dups \n",
    "            end\n",
    "\n",
    "            push!(content, Gene(id, dna, rev))\n",
    "        end \n",
    "    end    \n",
    "    \n",
    "    return Chromosome(content)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0db29f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genome\n",
    "\n",
    "mutable struct Genome\n",
    "    data::Vector{Chromosome}\n",
    "    id_to_char::Dict{Int, Char}\n",
    "    char_to_id::Dict{Char, Int}\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Genome, b::Genome)\n",
    "    return a.data == a.data && a.id_to_char == b.id_to_char && a.char_to_id == b.char_to_id\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ef463477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "string_to_genome (generic function with 1 method)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# str -> genome \n",
    "\n",
    "function string_to_genome(s, id_counter, id_to_char, char_to_id, target)\n",
    "     \"\"\"\n",
    "    Converts string to type Genome \n",
    "    \"\"\"\n",
    "\n",
    "    chrom_list_str = Vector{String}()\n",
    "\n",
    "    chrom = \"\"\n",
    "    new_chrom = true \n",
    "    linear = false\n",
    "\n",
    "    for (i, g) in pairs(s)\n",
    "        # creating a new chromosome \n",
    "        if new_chrom \n",
    "            linear = g == '.'\n",
    "            chrom = chrom * g\n",
    "            new_chrom = false  \n",
    "            if i == length(s) && !linear # single gene chrom \n",
    "                throw(ArgumentError(\"Circular chromosome must start and end with the same gene (case-sensitive).\"))\n",
    "            end \n",
    "        # appending more genes to existing chromosome \n",
    "        else  \n",
    "            if linear  \n",
    "                chrom = chrom * g \n",
    "                # close linear chrom (include telo in chrom) \n",
    "                if g == '.' \n",
    "                    push!(chrom_list_str, chrom) \n",
    "                    chrom = \"\"\n",
    "                    new_chrom = true \n",
    "                elseif i == length(s) \n",
    "                    throw(ArgumentError(\"Linear chromosome must start and end with telomeres.\"))\n",
    "                end \n",
    "            else  # circular\n",
    "                if g == chrom[1]  # close circular chrom \n",
    "                    push!(chrom_list_str, chrom)\n",
    "                    chrom = \"\" \n",
    "                    new_chrom = true \n",
    "                elseif i == length(s) \n",
    "                    throw(ArgumentError(\"Circular chromosome must start and end with the same gene (case-sensitive).\"))\n",
    "                else \n",
    "                    chrom = chrom * g\n",
    "                end \n",
    "            end \n",
    "        end \n",
    "    end\n",
    "\n",
    "    # convert chrom_list_str (list of strings) to chrom_list\n",
    "    chrom_list = Chromosome[]\n",
    "    \n",
    "    for c_str in chrom_list_str\n",
    "        c = Chromosome(c_str, id_counter, id_to_char, char_to_id, target)\n",
    "        push!(chrom_list, c)\n",
    "    end \n",
    "    \n",
    "    return Genome(chrom_list, id_to_char, char_to_id)\n",
    "end \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9b91ea5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene end (used in adj)\n",
    "\n",
    "@with_kw struct GeneEnd\n",
    "    gene::AbstractGene\n",
    "    head::Bool=true\n",
    "end\n",
    "\n",
    "function GeneEnd(T::Telomere)\n",
    "    return GeneEnd(Telomere(), false)  # head value doesn't matter   \n",
    "end \n",
    "\n",
    "function Base.:(==)(a::GeneEnd, b::GeneEnd)\n",
    "    # print(\"comparing gene eneds:: \")\n",
    "    # show(a.gene)\n",
    "    # show(b.gene)\n",
    "    # print(\"::\")\n",
    "    if a.gene == Telomere() && b.gene == Telomere() \n",
    "        return true \n",
    "    elseif a.gene == Telomere() || b.gene == Telomere() \n",
    "        return false \n",
    "    else \n",
    "        return a.gene.id == b.gene.id && a.head == b.head\n",
    "    end \n",
    "end\n",
    "\n",
    "function Base.show(ge::GeneEnd)\n",
    "    if ge.gene == Telomere() \n",
    "        show(ge.gene)\n",
    "        return \n",
    "    end \n",
    "    \n",
    "    if ge.head\n",
    "        print(string(ge.gene.dna) *\":h\")\n",
    "    else\n",
    "        print(string(ge.gene.dna) * \":t\")\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4ef3f032",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjacency\n",
    "mutable struct Adjacency \n",
    "    left::GeneEnd\n",
    "    right::GeneEnd\n",
    "end\n",
    "\n",
    "function Base.:(==)(a::Adjacency, b::Adjacency)\n",
    "     # order of the gene ends doesn't matter\n",
    "    if a.left == b.left && a.right == b.right || a.left == b.right && a.right == b.left  \n",
    "        return true \n",
    "    end \n",
    "    return false \n",
    "end \n",
    "\n",
    "function Base.show(adj::Adjacency)\n",
    "    left = \"\"\n",
    "    right = \"\"\n",
    "    \n",
    "    if adj.left.gene == Telomere() \n",
    "        left = \".\"\n",
    "    else \n",
    "        left_gene = adj.left\n",
    "        (left_gene.head == true) ? left = left_gene.gene.dna * \":h\" : left = left_gene.gene.dna * \":t\" \n",
    "    end\n",
    "    \n",
    "    if adj.right.gene == Telomere() \n",
    "        right = \".\"\n",
    "    else \n",
    "        right_gene = adj.right\n",
    "        (right_gene.head == true) ? right = right_gene.gene.dna * \":h\" : right = right_gene.gene.dna * \":t\"\n",
    "    end \n",
    "\n",
    "    print(\"(\", left, \",\", right, \")\")\n",
    "end \n",
    "\n",
    "\n",
    "# logging \n",
    "\n",
    "function Base.show(adj::Adjacency, blue::Bool)\n",
    "    left = \"\"\n",
    "    right = \"\"\n",
    "    \n",
    "    if adj.left.gene == Telomere() \n",
    "        left = \".\"\n",
    "    else \n",
    "        left_gene = adj.left\n",
    "        (left_gene.head == true) ? left = left_gene.gene.dna * \":h\" : left = left_gene.gene.dna * \":t\" \n",
    "    end\n",
    "    \n",
    "    if adj.right.gene == Telomere() \n",
    "        right = \".\"\n",
    "    else \n",
    "        right_gene = adj.right\n",
    "        (right_gene.head == true) ? right = right_gene.gene.dna * \":h\" : right = right_gene.gene.dna * \":t\"\n",
    "    end \n",
    "\n",
    "    s = \"(\" * left * \",\" * right * \")\"\n",
    "    \n",
    "    if blue \n",
    "        printstyled(s; color = :blue)\n",
    "    else\n",
    "        printstyled(s; color = :green) \n",
    "    end \n",
    "\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cb7a36eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# showing/printing collections of adj\n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency})\n",
    "    for adj in adj_list \n",
    "        show(adj) \n",
    "    end \n",
    "end \n",
    "\n",
    "function Base.show(adj_set::Set{Adjacency})\n",
    "    for adj in adj_set \n",
    "        show(adj) \n",
    "    end \n",
    "end \n",
    "\n",
    "function Base.show(adj_set::Set{Adjacency})\n",
    "    for adj in adj_set \n",
    "        show(adj) \n",
    "    end \n",
    "end \n",
    "\n",
    "# logging\n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency}, changed_adj1::Adjacency, changed_adj2::Adjacency)\n",
    "    for adj in adj_list \n",
    "        if adj == changed_adj1 || adj == changed_adj2\n",
    "            show(adj, true)\n",
    "        else \n",
    "            show(adj) \n",
    "        end \n",
    "    end \n",
    "end \n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency}, changed_adj::Adjacency) \n",
    "    for adj in adj_list \n",
    "        if adj == changed_adj\n",
    "            show(adj, true)\n",
    "        else \n",
    "            show(adj) \n",
    "        end \n",
    "    end \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f29ed684",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_to_adj_listset (generic function with 1 method)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function genome_to_adj_listset_helper(linear::Bool, chrom::Chromosome, genes::Vector{AbstractGene}, adj_list::Vector{Adjacency}, adj_set::Set{Adjacency})\n",
    "    # first gene end \n",
    "    if linear \n",
    "        left = GeneEnd(Telomere())\n",
    "        end_idx = length(genes) - 1\n",
    "    else \n",
    "        left = GeneEnd(genes[1], !genes[1].reverse)\n",
    "        end_idx = length(genes)\n",
    "    end  \n",
    "\n",
    "    for i in 2:end_idx\n",
    "        gene = genes[i]\n",
    "        reversed = gene.reverse  \n",
    "        \n",
    "        right = GeneEnd(gene, reversed)  # if reversed, right_end_adj is head (& left_end_adj is tail) \n",
    "        adj = Adjacency(left, right)\n",
    "        push!(adj_list, adj)\n",
    "        push!(adj_set, adj)\n",
    "\n",
    "        left = GeneEnd(gene, !reversed)\n",
    "    end \n",
    "\n",
    "    # last gene end \n",
    "    if linear \n",
    "        right = GeneEnd(Telomere())\n",
    "    else \n",
    "        right = GeneEnd(genes[1], genes[1].reverse)\n",
    "    end  \n",
    "\n",
    "    adj = Adjacency(left, right)\n",
    "    push!(adj_list, adj)\n",
    "    push!(adj_set, adj)\n",
    "end \n",
    "\n",
    "function genome_to_adj_listset(genome::Genome)\n",
    "    adj_list = Vector{Adjacency}()\n",
    "    adj_set = Set{Adjacency}()\n",
    "\n",
    "    for chrom in genome.data\n",
    "        genes = chrom.genes\n",
    "        if genes[1] == Telomere()  # linear \n",
    "            genome_to_adj_listset_helper(true, chrom, genes, adj_list, adj_set)\n",
    "        else  # circular \n",
    "            genome_to_adj_listset_helper(false, chrom, genes, adj_list, adj_set)\n",
    "        end\n",
    "    end \n",
    "    \n",
    "    return adj_list, adj_set\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e0b419e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AdjacencyGraph"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adjacency graph \n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    num_dcj_ops::Int\n",
    "end\n",
    "\n",
    "function AdjacencyGraph(src::Genome, target::Genome)\n",
    "    # process genomes (not explicitly creating the adj graph)    \n",
    "    src_adj_list, src_adj_set = genome_to_adj_listset(src)        \n",
    "    target_adj_list, target_adj_set = genome_to_adj_listset(target) \n",
    "    \n",
    "    print(\"SRC ADJ LIST\\n\")\n",
    "    show(src_adj_list)\n",
    "    # print(\"\\nSRC ADJ SET\\n\")\n",
    "    # show(src_adj_set)\n",
    "    \n",
    "    print(\"\\n|\\nv\\nTARGET ADJ LIST\\n\")\n",
    "    show(target_adj_list)\n",
    "    # print(\"\\nTARGET ADJ SET\\n\")\n",
    "    # show(target_adj_set)\n",
    "    \n",
    "    print(\"\\n\\n*************\\n\")\n",
    "    \n",
    "    src_geneid_to_location = process_adj_list(src_adj_list)   \n",
    "    target_geneid_to_location = process_adj_list(target_adj_list)\n",
    "    \n",
    "    # print(\"SRC GENE --> LOCATION DICT\\n\")\n",
    "    # show(src_geneid_to_location, src.id_to_char)  \n",
    "    # print(\"\\n\\n\", \"TARGET GENE --> LOCATION DICT\\n\")\n",
    "    # show(target_geneid_to_location, src.id_to_char)\n",
    "    # print(\"\\n*************\\n\\n\")\n",
    "\n",
    "    if src.id_to_char != target.id_to_char\n",
    "        throw(ArgumentError(\"ERROR: source and target genomes should have the same ID-->char dict\"))\n",
    "    end \n",
    "\n",
    "    # use adj graph representation to find dcj distance and operations \n",
    "    num_dcj_ops = find_dcj_dist_ops(src_adj_list, target_adj_list, src_geneid_to_location, target_geneid_to_location, src_adj_set, target_adj_set) \n",
    "    \n",
    "    return AdjacencyGraph(num_dcj_ops)\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "53b5edbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "calculate_distance (generic function with 1 method)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "assumptions \n",
    "genomes A and B are balanced (same letters of the same multiplicity; telomeres don't matter) \n",
    "no empty chromosomes (2 telomeres ..)\n",
    "\"\"\"\n",
    "\n",
    "# id_to_char = Dict{Int, Char}()  # TODO \n",
    "\n",
    "function calculate_distance(src::String, target::String)\n",
    "    # check_conditions(src, target)\n",
    "\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}()\n",
    "\n",
    "    target_genome = string_to_genome(target, id_counter, id_to_char, char_to_id, true)\n",
    "    src_genome = string_to_genome(src, id_counter, id_to_char, char_to_id, false)\n",
    "\n",
    "    ag = AdjacencyGraph(src_genome, target_genome)\n",
    "\n",
    "    return ag.num_dcj_ops\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "337313a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "reset_documentation (generic function with 1 method)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function reset_documentation(id_counter::Ref{Int}, id_to_char::Dict{Int, Char}, char_to_id::Dict{Char, Int})\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}() \n",
    "end \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
