{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_adj (generic function with 1 method)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using NBInclude\n",
    "@nbinclude(\"logging.ipynb\")\n",
    "@nbinclude(\"datastructs.ipynb\")\n",
    "@nbinclude(\"helpers.ipynb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "update_documentation (generic function with 1 method)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update documentation after a dcj operation\n",
    "\n",
    "function reassign_locs(pq::Adjacency, excluding_pq::Adjacency, u_idx::Int, v_idx::Int, gid_to_l::DefaultDict{Int, Vector{Int}})\n",
    "    # want to update dictionary given: \n",
    "    # src_adjs[u_idx] = pq \n",
    "    # src_adjs[v_idx] = excluding_pq \n",
    "\n",
    "    idx = Ref{Int}(u_idx)\n",
    "    if pq.left.gene != Telomere() \n",
    "        assign_ge_idx_to_gid_to_locdict(pq.left, idx, gid_to_l)\n",
    "    end \n",
    "    if pq.right != Telomere() \n",
    "        assign_ge_idx_to_gid_to_locdict(pq.right, idx, gid_to_l)\n",
    "    end \n",
    "    \n",
    "    idx = Ref{Int}(v_idx)\n",
    "    if excluding_pq.left.gene != Telomere() \n",
    "        assign_ge_idx_to_gid_to_locdict(excluding_pq.left, idx, gid_to_l)\n",
    "    end \n",
    "    if excluding_pq.right.gene != Telomere() \n",
    "        assign_ge_idx_to_gid_to_locdict(excluding_pq.right, idx, gid_to_l) \n",
    "    end \n",
    "end \n",
    "\n",
    "function update_src_adjs(p_pq::Adjacency, excluding_p_pq::Adjacency, u_idx::Int, v_idx::Int, src_adjs::Vector{Adjacency})\n",
    "    if u_idx == 0  # created a new adjacency \n",
    "        u_idx = length(src_adjs)+1\n",
    "        push!(src_adjs, p_pq)\n",
    "    end\n",
    "    if v_idx == 0  # created a new adjacency \n",
    "        v_idx = length(src_adjs)+1\n",
    "        push!(src_adjs, p_pq)\n",
    "    end \n",
    "    \n",
    "    src_adjs[u_idx] = p_pq \n",
    "    src_adjs[v_idx] = excluding_p_pq \n",
    "\n",
    "    return u_idx, v_idx\n",
    "end \n",
    "\n",
    "function update_documentation(p::GeneEnd, q::GeneEnd, u_idx::Int, v_idx::Int, src_adjs::Vector{Adjacency}, src_gid_to_l::DefaultDict{Int, Vector{Int}})\n",
    "    if q == GeneEnd(Telomere())\n",
    "        #  replace u in A by...\n",
    "        # {p} and (u\\{p})  \n",
    "        p_adj = Adjacency(p, GeneEnd(Telomere()))\n",
    "        excluding_p = Adjacency(other_adjacency_end(p, src_adjs[u_idx]), GeneEnd(Telomere()))\n",
    "    \n",
    "        # p, excluding_p\n",
    "        p_pq  = p_adj\n",
    "        excluding_p_pq = excluding_p\n",
    "    else \n",
    "        #  replace adj u and v in A by...\n",
    "        # {p, q} and \n",
    "        pq = Adjacency(p, q)\n",
    "    \n",
    "        # u\\{p}) U (v\\{q}\n",
    "        other_ge_u = other_adjacency_end(p, src_adjs[u_idx])\n",
    "        other_ge_v = other_adjacency_end(q, src_adjs[v_idx])\n",
    "        excluding_pq = Adjacency(other_ge_u, other_ge_v) \n",
    "\n",
    "        # pq, excluding_pq\n",
    "        p_pq  = pq\n",
    "        excluding_p_pq = excluding_pq\n",
    "    end \n",
    "\n",
    "\n",
    "    # update src adjs\n",
    "    u_idx, v_idx = update_src_adjs(p_pq, excluding_p_pq, u_idx, v_idx, src_adjs)\n",
    "    # update geneend locations  \n",
    "    reassign_locs(p_pq, excluding_p_pq, u_idx, v_idx, src_gid_to_l)\n",
    "\n",
    "    return p_pq, excluding_p_pq\n",
    "end \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_dcj_dist_ops (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find dcj operations and distance\n",
    "\n",
    "function find_dcj_dist_ops(src_adjs::Vector{Adjacency}, tar_adj_list::Vector{Adjacency}, src_gid_to_l::DefaultDict{Int, Vector{Int}}, tar_gid_to_l::DefaultDict{Int, Vector{Int}}, mode::Int, id_to_str::Dict{Int, String})\n",
    "    count = 0  \n",
    "    telomere_idxs = Vector{Int}()\n",
    "    updated_adj_list = src_adjs\n",
    "\n",
    "    if mode >= 2\n",
    "        println(\"indexing through target adj list...\\n\") \n",
    "        println(\"processing adj with two gene ends... \")\n",
    "    end \n",
    "\n",
    "    # for each adj {p, q} in target genome \n",
    "    for (i, adj) in pairs(tar_adj_list) \n",
    "        p = adj.left  \n",
    "        q = adj.right  \n",
    "        \n",
    "        if p.gene == Telomere() || q.gene == Telomere()  # telomeres handled in next loop\n",
    "            push!(telomere_idxs, i)  \n",
    "\n",
    "        else \n",
    "            if mode >= 2\n",
    "                print(\"\\nADJ \", i, \" :::: \")\n",
    "                show(adj, false)\n",
    "                print(\"\\n\")\n",
    "            end \n",
    "            if mode >= 3\n",
    "                print(\"current source adjacencies :: \")\n",
    "                show(updated_adj_list)\n",
    "                println()\n",
    "            end \n",
    "            \n",
    "            #  u = ge of adj from genome A that == p\n",
    "            #  v = ge of adj from genome A that == q\n",
    "            u_idx = find_adj(p, src_gid_to_l, src_adjs)  \n",
    "            v_idx = find_adj(q, src_gid_to_l, src_adjs)  \n",
    "\n",
    "            #  if u != v, replace u and v in A by {p, q} and (u\\{p}) U (v\\{q})\n",
    "            if u_idx != v_idx  \n",
    "                adj_to_change1 = src_adjs[u_idx]\n",
    "                adj_to_change2  = src_adjs[v_idx]\n",
    "\n",
    "                if mode >= 3\n",
    "                    prev_adj_list = copy(updated_adj_list)\n",
    "                end\n",
    "\n",
    "                pq, excluding_pq = update_documentation(p, q, u_idx, v_idx, src_adjs, src_gid_to_l)\n",
    "                \n",
    "                if mode >= 1\n",
    "                    print(\"DCJ :: \")                    \n",
    "                    show(adj_to_change1, true)\n",
    "                    show(adj_to_change2, true)\n",
    "\n",
    "                    print(\" --> \")\n",
    "                    \n",
    "                    show(pq, true) \n",
    "                    show(excluding_pq, true)\n",
    "                    println()\n",
    "                end \n",
    "                \n",
    "                if mode == 2\n",
    "                    print(\"updated src adjs \")\n",
    "                    show(src_adjs)\n",
    "                    println()\n",
    "                end \n",
    "\n",
    "                if mode >= 3\n",
    "                    print(\"\\nprev adj list ::: \")\n",
    "                    show(prev_adj_list, adj_to_change1, adj_to_change2)\n",
    "                    print(\"\\nupdated adj list ::: \")\n",
    "                    updated_adj_list = reorder_adjs(src_adjs, src_gid_to_l)\n",
    "                    show(updated_adj_list, pq, excluding_pq)\n",
    "                    \n",
    "                    println(\"\\nupdated src_gid_to_location\")\n",
    "                    show(src_gid_to_l, id_to_str)\n",
    "                end \n",
    "                \n",
    "            \n",
    "                count += 1 \n",
    "            end   \n",
    "        end\n",
    "    end \n",
    "\n",
    "    if mode >= 2\n",
    "        println(\"\\nprocessing telomeres... \")\n",
    "    end \n",
    "\n",
    "    # for each telomere {p} in genome B \n",
    "    for idx in telomere_idxs\n",
    "        p = tar_adj_list[idx]\n",
    "        p_non_telo = other_adjacency_end(GeneEnd(Telomere()), p)\n",
    "        \n",
    "        if mode >= 2\n",
    "            print(\"\\nADJ \", idx, \" :::: \")\n",
    "            show(p, false)\n",
    "            print(\"\\n\")\n",
    "        end \n",
    "\n",
    "        if mode >= 3\n",
    "                print(\"current source adjacencies :: \")\n",
    "                show(updated_adj_list)\n",
    "                println()\n",
    "        end         \n",
    "\n",
    "        #  u = ge of adj from genome A that == p \n",
    "        u_idx = find_adj(p_non_telo, src_gid_to_l, src_adjs)  \n",
    "        \n",
    "        # #  if u is an adjacency, then replace u in A by {p} and (u\\{p}) \n",
    "        if other_adjacency_end(p_non_telo, src_adjs[u_idx]).gene != Telomere()\n",
    "            adj_to_change = src_adjs[u_idx]\n",
    "\n",
    "            if mode >= 3\n",
    "                prev_adj_list = copy(updated_adj_list)\n",
    "            end \n",
    "            \n",
    "            p, excluding_p = update_documentation(p_non_telo, GeneEnd(Telomere()), u_idx, 0, src_adjs, src_gid_to_l)\n",
    "            \n",
    "            if mode >= 1 \n",
    "                print(\"DCJ :: \")\n",
    "                show(adj_to_change, true)\n",
    "\n",
    "                print(\" --> \")\n",
    "            \n",
    "                show(p, true) \n",
    "                show(excluding_p, true)\n",
    "                println()\n",
    "            end \n",
    "\n",
    "            if mode >= 3\n",
    "                print(\"\\nprev adj list ::: \")\n",
    "                show(prev_adj_list, src_adjs[u_idx])\n",
    "                print(\"\\nupdated adj list ::: \")\n",
    "                updated_adj_list = reorder_adjs(src_adjs, src_gid_to_l)\n",
    "                show(updated_adj_list, p, excluding_p)\n",
    "                \n",
    "                println(\"\\nupdated src_gid_to_location\")\n",
    "                show(src_gid_to_l, id_to_str)\n",
    "            end \n",
    "            \n",
    "            if mode == 2 \n",
    "                print(\"updated src adjs \")\n",
    "                show(src_adjs)\n",
    "                println()\n",
    "            end \n",
    "\n",
    "            count += 1 \n",
    "        end \n",
    "    end \n",
    "\n",
    "    return count\n",
    "end \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "calculate_distance (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "finds the minimum dcj distance to get from the source genome --> target genome \n",
    "O(N) time, O(N) space where n = number of genes in genome\n",
    "\n",
    "src & target \n",
    "    source and target genomes (balanced = have the same genes/letters of the same multiplicity; telomeres don't matter)\n",
    "    \n",
    "    genome = list of chromosome(s)\n",
    "        chromosomes are separated by a comma (no spaces)\n",
    "    chromosome = list of genes \n",
    "        chromosomes can be circular (no periods) or linear (sandwiched between two periods)\n",
    "    gene = string with non-duplicated letter(s) (letter == gene), followed by its duplicate value in the whole genome (0 if there are no duplicates)\n",
    "        letters may be capitalized to denote reversed orientation of the gene (which strand of the DNA double helix it's on)\n",
    "        \n",
    "    e.g., \n",
    "        genome a0b1c0b0,b2\n",
    "        chromosome [a0 b1 c0 b0] [b2]\n",
    "        genes \"a0\" \"b1\" \"c0\" \"b0\" \"b2\"\n",
    "\n",
    "mode \n",
    "    \"none\" - doesn't print   \n",
    "    \"info\" - informative   \n",
    "    \"debug\" - prints information useful for debugging \n",
    "    \"trace\" - prints everything --- note: makes algo O(n^2) \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "function calculate_distance(src::String, target::String, mode::String)\n",
    "    if mode == \"info\"\n",
    "        mode_int = 1\n",
    "    elseif mode == \"debug\"\n",
    "        mode_int = 2 \n",
    "    elseif mode == \"trace\"\n",
    "        mode_int = 3 \n",
    "    else \n",
    "        mode_int = 0 \n",
    "    end \n",
    "    \n",
    "    # check_conditions(src, target)\n",
    "\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_str = Dict{Int, String}()\n",
    "    str_to_id = Dict{String, Int}()\n",
    "\n",
    "    target_genome = string_to_genome(target, id_counter, id_to_str, str_to_id, true)\n",
    "    src_genome = string_to_genome(src, id_counter, id_to_str, str_to_id, false)\n",
    "    \n",
    "    ag = AdjacencyGraph(src_genome, target_genome, mode_int)\n",
    "\n",
    "    return ag.num_dcj_ops\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# src = \".a0.,.b0.,.c0.,.d0.\"\n",
    "# target=\"A0B0C0D0\"\n",
    "# mode = \"info\"\n",
    "# calculate_distance(src, target, mode)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
