{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_to_adj_listset (generic function with 1 method)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using NBInclude\n",
    "@nbinclude(\"dcj_datastructs.ipynb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "process_adj_list (generic function with 1 method)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process adj list \n",
    "# create dictionary of gene ID --> index/location of gene's head and tail \n",
    "\n",
    "function process_adj_list_helper(ge::GeneEnd, idx::Ref{Int}, gid_to_loc::DefaultDict{Int, Vector{Int}})\n",
    "    if ge.gene == Telomere() \n",
    "        return\n",
    "    end  \n",
    "\n",
    "    if ge.head == true\n",
    "        gid_to_loc[ge.gene.id][2] = idx[]\n",
    "    else \n",
    "        gid_to_loc[ge.gene.id][1] = idx[]\n",
    "    end\n",
    "end \n",
    "\n",
    "function process_adj_list(adj_list:: Vector{Adjacency})\n",
    "    geneid_to_location = DefaultDict{Int, Vector{Int}}(() -> zeros(Int, 2)) # tail = idx 1, head = 2 in array\n",
    "    idx = Ref{Int}(1)\n",
    "\n",
    "    for adj in adj_list\n",
    "        process_adj_list_helper(adj.left, idx, geneid_to_location)\n",
    "        process_adj_list_helper(adj.right, idx, geneid_to_location)\n",
    "        idx[] += 1\n",
    "    end \n",
    "    \n",
    "    return geneid_to_location \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "reassign_locs (generic function with 1 method)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# helpers for dcj operations and distance\n",
    "\n",
    "# finds target gene end in source adjacency list \n",
    "# returns index and left/right in the source adj list\n",
    "function find_tar_ge_in_src_adjlist(target_ge::GeneEnd, src_gid_to_l::DefaultDict{Int, Vector{Int}}, tar_gid_to_l::DefaultDict{Int, Vector{Int}}, src_adj_list::Vector{Adjacency})\n",
    "    (target_ge.head == true) ? th_idx = 2 : th_idx = 1\n",
    "    # (left_tar_ge == true) ? tar_lr = 'L' : tar_lr = 'R'\n",
    "        \n",
    "    gene_id = target_ge.gene.id\n",
    "\n",
    "    src_ge_idx = src_gid_to_l[gene_id][th_idx]\n",
    "\n",
    "    adj = src_adj_list[src_ge_idx]\n",
    "    if adj.left.gene.id == gene_id\n",
    "        src_ge = adj.left\n",
    "        # src_lr = 'L'\n",
    "    else \n",
    "        src_ge = adj.right \n",
    "        # src_lr = 'R' \n",
    "    end \n",
    "\n",
    "    # print(\"\\n   GENEEND: \")\n",
    "    # show(target_ge)\n",
    "    # print(\"\\n     src genome idx: \", src_ge_idx, \" \", src_lr)\n",
    "    # print(\"\\n     target genome idx: \", tar_gid_to_l[gene_id][th_idx], \" \", tar_lr, \"\\n\")\n",
    "\n",
    "    return [src_ge, src_ge_idx]\n",
    "end \n",
    "\n",
    "function other_adjacency_end(ge::GeneEnd, adj::Adjacency)\n",
    "    if adj.left == ge\n",
    "        return adj.right\n",
    "    else\n",
    "        return adj.left\n",
    "    end \n",
    "end\n",
    "\n",
    "function non_telo_end(adj::Adjacency)\n",
    "    if adj.left.gene == Telomere()\n",
    "        return adj.right\n",
    "    else\n",
    "        return adj.left\n",
    "    end \n",
    "end\n",
    "\n",
    "function combine_ge(u::GeneEnd, u_idx::Int, v::GeneEnd, v_idx::Int)\n",
    "    if u_idx < v_idx \n",
    "        u_lt_v = true \n",
    "    else \n",
    "        u_lt_v = false \n",
    "    end \n",
    "\n",
    "    if !u_lt_v \n",
    "        return Adjacency(v, u)\n",
    "    else \n",
    "        return Adjacency(u, v)\n",
    "    end\n",
    "\n",
    "end \n",
    "\n",
    "function reassign_locs(gid_to_l, adj_list, u_idx, v_idx)\n",
    "    for (i, adj) in pairs(view(adj_list, min(u_idx, v_idx):max(u_idx, v_idx)))\n",
    "        left = adj.left \n",
    "        right = adj.right \n",
    "\n",
    "        idx = Ref{Int}(i + min(u_idx, v_idx) - 1)\n",
    "\n",
    "        process_adj_list_helper(left, idx, gid_to_l)\n",
    "        process_adj_list_helper(right, idx, gid_to_l)\n",
    "    end \n",
    "end \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_dcj_dist_ops (generic function with 1 method)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function update_adj_set(p::GeneEnd, q::GeneEnd, u::GeneEnd, v::GeneEnd, u_idx::Int, v_idx::Int, src_adj_list::Vector{Adjacency}, src_adj_set::Set{Adjacency}) \n",
    "    #  replace adj u and v in A by ( {p, q} and (u\\{p}) U (v\\{q}) )\n",
    "    pq = combine_ge(p, u_idx, q, v_idx)\n",
    "\n",
    "    other_ge_u = other_adjacency_end(u, src_adj_list[u_idx])\n",
    "    other_ge_v = other_adjacency_end(v, src_adj_list[v_idx])\n",
    "    excluding_pq = combine_ge(other_ge_u, u_idx, other_ge_v, v_idx) \n",
    "    \n",
    "    # print(\"******exlcuding pq\")\n",
    "    # show(excluding_pq)\n",
    "    # print(\"********\")\n",
    "\n",
    "    #  remove adj with u, v in src_adj_set \n",
    "    delete!(src_adj_set, src_adj_list[u_idx])\n",
    "    delete!(src_adj_set, src_adj_list[v_idx])\n",
    "    \n",
    "    #  add adjusted adjacencies to src_adj_set\n",
    "    push!(src_adj_set, pq)\n",
    "    push!(src_adj_set, excluding_pq)\n",
    "\n",
    "    # print(\"  src_adj_set:\")\n",
    "    # show(src_adj_set)\n",
    "    # print(\"\\n\")\n",
    "end \n",
    "\n",
    "function update_adj_list_gid_loc_dict() \n",
    "    # update adj lists \n",
    "        print(\"\\nDCJ Operation (Cycle) :: \")\n",
    "        show(src_adj_list, u_idx, v_idx)\n",
    "        \n",
    "        print(\"\\n\")\n",
    "        show(pq)\n",
    "        show(excluding_pq)\n",
    "        if u_idx < v_idx \n",
    "            src_adj_list[u_idx] = pq\n",
    "            src_adj_list[v_idx] = excluding_pq\n",
    "            reverse!(view(src_adj_list, u_idx+1:v_idx-1))\n",
    "        else \n",
    "            src_adj_list[u_idx] = excluding_pq\n",
    "            src_adj_list[v_idx] = pq\n",
    "            reverse!(view(src_adj_list, v_idx+1:u_idx-1))\n",
    "        end  \n",
    "    \n",
    "        # update gid-->loc dicts \n",
    "        reassign_locs(src_gid_to_l, src_adj_list, u_idx, v_idx)\n",
    "end \n",
    "\n",
    "# find dcj operations and distance\n",
    "\n",
    "function find_dcj_dist_ops(src_adj_list::Vector{Adjacency}, tar_adj_list::Vector{Adjacency}, src_gid_to_l::DefaultDict{Int, Vector{Int}}, tar_gid_to_l::DefaultDict{Int, Vector{Int}}, src_adj_set::Set{Adjacency}, target_adj_set::Set{Adjacency})\n",
    "    count = 0  \n",
    "    # cycles = 0 \n",
    "    # odd_paths = 0 \n",
    "    telomere_idxs = Vector{Int}\n",
    "\n",
    "    print(\"indexing through target adj list...\")\n",
    "    \n",
    "    # for each adj {p, q} in target genome \n",
    "    for (i, adj) in pairs(tar_adj_list) \n",
    "        print(\"\\n\\n\")\n",
    "        show(src_adj_list)\n",
    "         \n",
    "        print(\"\\n\\nADJ \", i, \" :::: \")\n",
    "        show(adj)\n",
    "        print(\"\\n\")\n",
    "\n",
    "        p = adj.left  \n",
    "        q = adj.right  \n",
    "\n",
    "        if p.gene == Telomere() || q.gene == Telomere()  # telomeres handled in next loop\n",
    "            push!(telomere_idxs, i)  \n",
    "        end \n",
    "\n",
    "        #  let u be element of genome A that contains p \n",
    "        #  let v be element of genome A that contains q \n",
    "        u, u_idx = find_tar_ge_in_src_adjlist(p, src_gid_to_l, tar_gid_to_l, src_adj_list)  \n",
    "        v, v_idx = find_tar_ge_in_src_adjlist(q, src_gid_to_l, tar_gid_to_l, src_adj_list)  \n",
    "        \n",
    "        #  if u != v, replace u and v in A by {p, q} and (u\\{p}) U (v\\{q})\n",
    "        if u_idx != v_idx             \n",
    "            update_adj_set(p, q, u, v, u_idx, v_idx, src_adj_list, src_adj_set)\n",
    "           \n",
    "            print(\"updated adj set ::: \")\n",
    "            show(src_adj_set)\n",
    "            print(\"\\nupdated adj list ::: \")\n",
    "            print_adj_set_as_adj_list(src_adj_set)\n",
    "\n",
    "            #  update_adj_list_gid_loc_dict()\n",
    "            \n",
    "            count += 1 \n",
    "        end   \n",
    "    end \n",
    "\n",
    "    # for each telomere {p} in genome B \n",
    "    # for idx in telomere_idxs\n",
    "    #     telo_adj = tar_adj_list[idx]\n",
    "    #     telo_ge = other_adjacency_end(telo_adj, Telomere())\n",
    "\n",
    "    #     #  let u be element of genome A that contains p \n",
    "    #     u, u_idx = find_tar_ge_in_src_adjlist(p, src_gid_to_l, tar_gid_to_l, src_adj_list)  \n",
    "\n",
    "    #     #  if u is an adjacency, then replace u in A by {p} and (u\\{p})\n",
    "    #     if other_adjacency_end(tar_adj_list[u_idx], u) != Telomere() \n",
    "    #         # update_adj_set(p, u, u_idx,src_adj_list, src_adj_set)\n",
    "    #         # update_adj_list_gid_loc_dict()\n",
    "\n",
    "    #         count += 1 \n",
    "    #     end \n",
    "    # end \n",
    "\n",
    "    return count \n",
    "    # return cycles, odd_paths\n",
    "end \n",
    "\n",
    "# src = \"abc\"\n",
    "# target = \"cba\"\n",
    "\n",
    "# calculate_distance(src, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRC ADJ LIST\n",
      "(.,a:t)(a:h,.)(b:h,c:t)(c:h,b:t)\n",
      "|\n",
      "v\n",
      "TARGET ADJ LIST\n",
      "(c:h,b:t)(b:h,c:t)(.,a:t)(a:h,.)\n",
      "\n",
      "*************\n",
      "indexing through target adj list...\n",
      "\n",
      "(.,a:t)(a:h,.)(b:h,c:t)(c:h,b:t)\n",
      "\n",
      "ADJ 1 :::: (c:h,b:t)\n",
      "\n",
      "\n",
      "(.,a:t)(a:h,.)(b:h,c:t)(c:h,b:t)\n",
      "\n",
      "ADJ 2 :::: (b:h,c:t)\n",
      "\n",
      "\n",
      "(.,a:t)(a:h,.)(b:h,c:t)(c:h,b:t)\n",
      "\n",
      "ADJ 3 :::: (.,a:t)\n"
     ]
    },
    {
     "ename": "LoadError",
     "evalue": "MethodError: no method matching push!(::Type{Vector{Int64}}, ::Int64)\n\n\u001b[0mClosest candidates are:\n\u001b[0m  push!(::Any, ::Any, \u001b[91m::Any\u001b[39m)\n\u001b[0m\u001b[90m   @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m\u001b[4mabstractarray.jl:3413\u001b[24m\u001b[39m\n\u001b[0m  push!(::Any, ::Any, \u001b[91m::Any\u001b[39m, \u001b[91m::Any...\u001b[39m)\n\u001b[0m\u001b[90m   @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m\u001b[4mabstractarray.jl:3414\u001b[24m\u001b[39m\n\u001b[0m  push!(\u001b[91m::Vector{Any}\u001b[39m, ::Any)\n\u001b[0m\u001b[90m   @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m\u001b[4marray.jl:1125\u001b[24m\u001b[39m\n\u001b[0m  ...\n",
     "output_type": "error",
     "traceback": [
      "MethodError: no method matching push!(::Type{Vector{Int64}}, ::Int64)\n\n\u001b[0mClosest candidates are:\n\u001b[0m  push!(::Any, ::Any, \u001b[91m::Any\u001b[39m)\n\u001b[0m\u001b[90m   @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m\u001b[4mabstractarray.jl:3413\u001b[24m\u001b[39m\n\u001b[0m  push!(::Any, ::Any, \u001b[91m::Any\u001b[39m, \u001b[91m::Any...\u001b[39m)\n\u001b[0m\u001b[90m   @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m\u001b[4mabstractarray.jl:3414\u001b[24m\u001b[39m\n\u001b[0m  push!(\u001b[91m::Vector{Any}\u001b[39m, ::Any)\n\u001b[0m\u001b[90m   @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m\u001b[4marray.jl:1125\u001b[24m\u001b[39m\n\u001b[0m  ...\n",
      "",
      "Stacktrace:",
      " [1] find_dcj_dist_ops(src_adj_list::Vector{Adjacency}, tar_adj_list::Vector{Adjacency}, src_gid_to_l::DefaultDict{Int64, Vector{Int64}, var\"#28#29\"}, tar_gid_to_l::DefaultDict{Int64, Vector{Int64}, var\"#28#29\"}, src_adj_set::Set{Adjacency}, target_adj_set::Set{Adjacency})",
      "   @ Main .\\In[15]:71",
      " [2] AdjacencyGraph(src::Genome, target::Genome)",
      "   @ Main .\\In[17]:49",
      " [3] calculate_distance(src::String, target::String)",
      "   @ Main .\\In[17]:70",
      " [4] top-level scope",
      "   @ In[29]:98"
     ]
    }
   ],
   "source": [
    "# given ge we're looking for \n",
    "# finds new ge that corresponds to given ge (tail/head) \n",
    "# returns new ge in new adjacency \n",
    "function find_next_adj(ge::GeneEnd, adj_set::Set{Adjacency}, adj_list::Vector{Adjacency}, processed_adjs::Set{Adjacency}, adj::Adjacency) \n",
    "    for adj in adj_set \n",
    "        if adj in processed_adjs\n",
    "            continue \n",
    "        end \n",
    "\n",
    "        if adj.left.gene == ge.gene\n",
    "            println(\"111111111111111111111111111111\")\n",
    "            push!(adj_list, adj)\n",
    "            push!(processed_adjs, adj)\n",
    "            ge = other_adjacency_end(adj.left, adj)\n",
    "\n",
    "            show(adj)\n",
    "            return ge \n",
    "        elseif adj.right.gene == ge.gene\n",
    "            println(\"22222222222222222222222\")\n",
    "            push!(adj_list, adj)\n",
    "            push!(processed_adjs, adj)\n",
    "            ge = other_adjacency_end(adj.right, adj)\n",
    "\n",
    "            show(adj)\n",
    "            return ge \n",
    "        end \n",
    "    end \n",
    "    println(\"3333333333333333333333333333333\")\n",
    "    # push!(adj_list, adj)\n",
    "    return other_adjacency_end(ge, adj)\n",
    "end \n",
    "\n",
    "# O(N^2)\n",
    "function print_adj_set_as_adj_list(og_adj_set::Set{Adjacency})\n",
    "    telos = Vector{Adjacency}() \n",
    "    adj_set = copy(og_adj_set)\n",
    "    adj_list = Vector{Adjacency}() \n",
    "\n",
    "    # handle linear chroms first \n",
    "    for adj in adj_set\n",
    "        # show(adj)\n",
    "        if adj.left.gene == Telomere() \n",
    "            push!(telos, adj)\n",
    "        end \n",
    "    end \n",
    "\n",
    "    for adj in telos \n",
    "        # ge = adj.right  \n",
    "        ge = non_telo_end(adj) \n",
    "        push!(adj_list, adj)\n",
    "        delete!(adj_set, adj)\n",
    "\n",
    "        while ge.gene != Telomere()  \n",
    "            show(ge)\n",
    "            ge = find_next_adj(ge, adj_set, adj_list)\n",
    "        end \n",
    "        show(ge)\n",
    "    end  \n",
    "    \n",
    "    # handle circular chroms \n",
    "    processed_adjs = Set{Adjacency}() \n",
    "    for adj in adj_set  # only circular left \n",
    "        # print(\"\\nprocessing :: \")\n",
    "        # show(adj)\n",
    "        # println()\n",
    "\n",
    "        if adj in processed_adjs\n",
    "            continue \n",
    "        end \n",
    "        push!(processed_adjs, adj)\n",
    "        push!(adj_list, adj)\n",
    "        \n",
    "        first_gene = adj.left.gene \n",
    "        ge = GeneEnd(adj.right.gene, !adj.right.head)  # ge that we're looking for\n",
    "        first = true       \n",
    "        \n",
    "        while ge.gene != first_gene || first == true  \n",
    "            # print(\"LOOKING FOR \")\n",
    "            # show(ge)\n",
    "            ge = find_next_adj(ge, adj_set, adj_list, processed_adjs, adj)\n",
    "            # println()\n",
    "            first = false \n",
    "        end \n",
    "        # show(ge)\n",
    "        # print(\"updated adj_list ---------\")\n",
    "        # show(adj_list)    \n",
    "        # println()     \n",
    "    end \n",
    "    print(\"\\n**\")\n",
    "    show(adj_list)\n",
    "    print(\"*8\") \n",
    "end \n",
    "\n",
    "\n",
    "src = \".a.bc\"\n",
    "target = \"cb.a.\"\n",
    "\n",
    "calculate_distance(src, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "calculate_distance (generic function with 1 method)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adjacency graph \n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    cycles::Int\n",
    "    odd_paths::Int\n",
    "end\n",
    "\n",
    "function Base.show(dict::DefaultDict{Int, Vector{Int}}, id_to_char::Dict{Int, Char})\n",
    "    for (key, value) in dict\n",
    "        # println(\"$(key.dna) (id:$(key.id)) => $(value)\")\n",
    "        println(id_to_char[key], \" $(key) => $(value)\")\n",
    "    end\n",
    "end \n",
    "\n",
    "function AdjacencyGraph(src::Genome, target::Genome)\n",
    "\n",
    "    # process genomes (not explicitly creating the adj graph)    \n",
    "    src_adj_list, src_adj_set = genome_to_adj_listset(src)        \n",
    "    target_adj_list, target_adj_set = genome_to_adj_listset(target) \n",
    "    \n",
    "    print(\"SRC ADJ LIST\\n\")\n",
    "    show(src_adj_list)\n",
    "    # print(\"\\nSRC ADJ SET\\n\")\n",
    "    # show(src_adj_set)\n",
    "    print(\"\\n|\\nv\\nTARGET ADJ LIST\\n\")\n",
    "    show(target_adj_list)\n",
    "    # print(\"\\nTARGET ADJ SET\\n\")\n",
    "    # show(target_adj_set)\n",
    "    \n",
    "    print(\"\\n\\n*************\\n\")\n",
    "\n",
    "    # print_adj_set_as_adj_list(src_adj_set)\n",
    "    \n",
    "    src_geneid_to_location = process_adj_list(src_adj_list)   \n",
    "    target_geneid_to_location = process_adj_list(target_adj_list)\n",
    "    \n",
    "    if src.id_to_char != target.id_to_char\n",
    "        throw(ArgumentError(\"ERROR: Source ID to char dict is not equal to target's.\"))\n",
    "    end \n",
    "\n",
    "    # print(\"SRC GENE --> LOCATION DICT\\n\")\n",
    "    # show(src_geneid_to_location, src.id_to_char)  \n",
    "    # print(\"\\n\\n\", \"TARGET GENE --> LOCATION DICT\\n\")\n",
    "    # show(target_geneid_to_location, src.id_to_char)\n",
    "    # print(\"\\n*************\\n\\n\")\n",
    "\n",
    "    # use adj graph representation to find dcj distance and operations \n",
    "    # cycles, odd_paths = \n",
    "    find_dcj_dist_ops(src_adj_list, target_adj_list, src_geneid_to_location, target_geneid_to_location, src_adj_set, target_adj_set) \n",
    "    \n",
    "    # return AdjacencyGraph(cycles, odd_paths)\n",
    "end \n",
    "\n",
    "\"\"\"\n",
    "assumptions \n",
    "genomes A and B have the same letters of the same multiplicity; telomeres don't matter \n",
    "no empty chromosomes (2 telomeres ..)\n",
    "\"\"\"\n",
    "\n",
    "function calculate_distance(src::String, target::String)\n",
    "    # check_conditions(src, target)\n",
    "\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}()\n",
    "\n",
    "    target_genome = string_to_genome(target, id_counter, id_to_char, char_to_id, true)\n",
    "    src_genome = string_to_genome(src, id_counter, id_to_char, char_to_id, false)\n",
    "\n",
    "    ag = AdjacencyGraph(src_genome, target_genome)\n",
    "\n",
    "    # dcj_distance = src_genome.num_genes - ag.cycles - (ag.ab_paths / 2)  # or target_genome.num_genes\n",
    "\n",
    "    # return dcj_distance\n",
    "end\n",
    "\n",
    "# src = \"a.bcde.f\"\n",
    "# target = \"abcdef\"\n",
    "\n",
    "# calculate_distance(src, target)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
