{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " \"\"\"The abstraction of a chromosome in DCJ model, a list of telomeres and genes\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genes: list of genes or string (req)\n",
    "        The object to be converted into a chromosome.\n",
    "\n",
    "    \"\"\"\n",
    "mutable struct Chromosome\n",
    "    genes::Vector{AbstractGene}\n",
    "    # gene_set::Set{String}\n",
    "end\n",
    "\n",
    "# mutable struct ChromosomeStatic\n",
    "#     string_d::Dict{String, Int}  # gene to ID \n",
    "#     uid_counter::Int\n",
    "#     ChromosomeStatic() = new(Dict{String, Int}(), 0)\n",
    "# end\n",
    "\n",
    "# const chromosome_static = ChromosomeStatic()\n",
    "\n",
    "function Chromosome(genes::Vector{})\n",
    "    # gene_set = Set{AbstractGene}()\n",
    "    content = AbstractGene[]\n",
    "    \n",
    "    # genes = list of genes \n",
    "    if isa(genes, Vector{AbstractGene})\n",
    "        content = genes\n",
    "     \n",
    "    # throw(TypeError(\"Chromosome must be a list of Gene instances or a string of unique genes.\"))\n",
    "\n",
    "    # parse string \n",
    "    # else \n",
    "    #     genes_str = collect(genes)\n",
    "    #     if !((genes_str[1] == '.') == (genes_str[end] == '.'))\n",
    "    #         throw(ArgumentError(\"Linear Chromosome must start and end with telomeres.\"))\n",
    "    #     end\n",
    "    #     for (i, s) in enumerate(genes_str)\n",
    "    #         dna = string(s)\n",
    "    #         rev = isuppercase(s)\n",
    "    #         telomere = s == '.'\n",
    "    #         # handle gene \n",
    "    #         if s in gene_set && !telomere\n",
    "    #             throw(ArgumentError(\"Duplicated genes are not allowed. ($s)\"))\n",
    "    #         elseif s in keys(chromosome_static.string_d)\n",
    "    #             uid = chromosome_static.string_d[s]\n",
    "    #             push!(gene_set, s)\n",
    "    #         else\n",
    "    #             chromosome_static.string_d[s] = chromosome_static.uid_counter\n",
    "    #             push!(gene_set, s)\n",
    "    #             uid = chromosome_static.uid_counter\n",
    "    #             chromosome_static.uid_counter += 1\n",
    "    #         end\n",
    "    #         # handle telomere\n",
    "    #         if telomere && !(i == 1 || i == length(genes_str))\n",
    "    #             throw(ArgumentError(\"Telomere cannot appear in middle of chromosome.\"))\n",
    "    #         elseif telomere\n",
    "    #             push!(content, Telomere())\n",
    "    #         else\n",
    "    #             push!(content, Gene(uid, dna, rev))\n",
    "    #         end\n",
    "    #     end\n",
    "    end\n",
    "    return Chromosome(content, gene_set)\n",
    "end\n",
    "\n",
    "function clear_chromosome_static()\n",
    "    chromosome_static.string_d = Dict{String, Int}()\n",
    "    chromosome_static.uid_counter = 0\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### testing chromosome works \n",
    "\n",
    "# testing strings \n",
    "\n",
    "id_counter = Ref{Int}(1)\n",
    "id_to_char = Dict{Int, Char}()\n",
    "char_to_id = Dict{Char, Int}()\n",
    "\n",
    "Chromosome(\".AbcA.\", id_counter, id_to_char, char_to_id)\n",
    "\n",
    "# print(\"\\n\\n******************\\nid_counter: \", id_counter[], \"\\n\", \"id_to_char\", id_to_char, \"\\n\",\"char_to_id: \", char_to_id)\n",
    "\n",
    "\n",
    "id_counter = Ref{Int}(1)\n",
    "id_to_char = Dict{Int, Char}()\n",
    "char_to_id = Dict{Char, Int}()\n",
    "\n",
    "Chromosome(\".Abcd.\", id_counter, id_to_char, char_to_id, true)\n",
    "Chromosome(\".Adbc.\", id_counter, id_to_char, char_to_id, false)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing genes \n",
    "x = Telomere()\n",
    "y = Gene(1, \"A\", true)\n",
    "z = Gene(2, \"B\", true)\n",
    "v = Telomere()\n",
    "\n",
    "id_set = Set{Int}() \n",
    "\n",
    "Chromosome([y, y, z], id_set)  ## why doesn't it find the right constructor? \n",
    "\n",
    "# testing rand \n",
    "# x = 1 \n",
    "# Chromosome(x) ## why doesn't it throw error??? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function Chromosome(genes::Vector{AbstractGene}, id_set::Set{Int})\n",
    "    # print(\"entered Chromosome constructor for genes********************************************************\")\n",
    "    \n",
    "    for gene in genes\n",
    "        telomere = show(gene) == \".\"\n",
    "        if !telomere \n",
    "            if gene.id âˆ‰ id_set \n",
    "                push!(id_set, gene.id)\n",
    "            else \n",
    "                throw(ArgumentError(\"There are duplicate genes with the same ID in this genome.\"))\n",
    "            end\n",
    "        end \n",
    "    end \n",
    "    content = genes\n",
    "\n",
    "    return Chromosome(content)\n",
    "\n",
    "end \n",
    "\n",
    "\n",
    "\n",
    "function Chromosome(::Any) \n",
    "    throw(TypeError(\"Chromosome must be a list of Gene instances with unique IDs or a string of genes.\"))\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Funcs \n",
    "\n",
    "\"\"\"\n",
    "checks for correct conditions \n",
    "    - only letters and dots\n",
    "    - even number of telomeres in both genomes (can't have the start of a chromosome without closing it)  \n",
    "\"\"\"\n",
    "function check_conditions(A::String, B::String)\n",
    "    set_A = Dict{Char, Int64}\n",
    "    set_B = Set{Char}()\n",
    "    tel_count_A = 0 \n",
    "    tel_count_B = 0 \n",
    "\n",
    "    for i in 1:length(A)\n",
    "        if A[i] == '.'\n",
    "            tel_count_A += 1\n",
    "        elseif isletter(A[i])\n",
    "            push!(set_A, lowercase(A[i])) # or uppercase, just keep consistent\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome A has non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    for i in 1:length(B) \n",
    "        if B[i] == '.'\n",
    "            tel_count_B += 1\n",
    "        elseif  isletter(A[i])\n",
    "            push!(set_B, lowercase(B[i]))\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome B has non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    if mod(tel_count_A, 2) != 0 || mod(tel_count_B, 2) != 0  \n",
    "         throw(ArgumentError(\"Error: Telomere error\"))\n",
    "        return false \n",
    "    elseif set_A != set_B\n",
    "        throw(ArgumentError(\"Error: Sets of genes in the two given genomes don't match\"))\n",
    "        return false\n",
    "    end \n",
    "\n",
    "\n",
    "    return true \n",
    "end \n",
    "\n",
    "# correct_conditions(\".abdc.\", \"abCd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## tests telomeres adn genes \n",
    "\n",
    "using Test \n",
    "\n",
    "function test_telomere_type()\n",
    "    a = Telomere()\n",
    "    @test typeof(a) == Telomere\n",
    "end\n",
    "\n",
    "function test_forward_reverse_gene()\n",
    "    @test isequal(Gene(0, \"d\", true), Gene(0, \"d\", false))\n",
    "end\n",
    "\n",
    "function test_telomere_repr()\n",
    "    @test show(Telomere()) == \".\"\n",
    "end\n",
    "\n",
    "function test_reverse_equality()\n",
    "    @test !isequal(Gene(0, \"D\", true), Gene(1, \"D\", true))\n",
    "end\n",
    "\n",
    "@testset \"Gene Tests\" begin\n",
    "    test_telomere_type()\n",
    "    test_forward_reverse_gene()\n",
    "    test_telomere_repr()\n",
    "    test_reverse_equality()\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inside Chromosome\n",
    "\n",
    "# check if empty chromosome .. \n",
    "if g == '.' \n",
    "    chrom = chrom * g \n",
    "    push!(chrom_list, chrom)\n",
    "    chrom = \"\"\n",
    "    continue \n",
    "end "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
