{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Tests: \u001b[91m\u001b[1mError During Test\u001b[22m\u001b[39m at \u001b[39m\u001b[1mIn[2]:22\u001b[22m\n",
      "  Got exception outside of a @test\n",
      "  UndefVarError: `Telomere` not defined\n",
      "  Stacktrace:\n",
      "    [1] \u001b[0m\u001b[1mtest_telomere_type\u001b[22m\u001b[0m\u001b[1m(\u001b[22m\u001b[0m\u001b[1m)\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[35mMain\u001b[39m \u001b[90m.\\\u001b[39m\u001b[90m\u001b[4mIn[2]:6\u001b[24m\u001b[39m\n",
      "    [2] \u001b[0m\u001b[1mmacro expansion\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[90m\u001b[4mIn[2]:23\u001b[24m\u001b[39m\u001b[90m [inlined]\u001b[39m\n",
      "    [3] \u001b[0m\u001b[1mmacro expansion\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[90mC:\\Users\\mgnli\\.julia\\juliaup\\julia-1.10.3+0.x64.w64.mingw32\\share\\julia\\stdlib\\v1.10\\Test\\src\\\u001b[39m\u001b[90m\u001b[4mTest.jl:1577\u001b[24m\u001b[39m\u001b[90m [inlined]\u001b[39m\n",
      "    [4] top-level scope\n",
      "  \u001b[90m    @\u001b[39m \u001b[90m\u001b[4mIn[2]:23\u001b[24m\u001b[39m\n",
      "    [5] \u001b[0m\u001b[1meval\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[90m.\\\u001b[39m\u001b[90m\u001b[4mboot.jl:385\u001b[24m\u001b[39m\u001b[90m [inlined]\u001b[39m\n",
      "    [6] \u001b[0m\u001b[1minclude_string\u001b[22m\u001b[0m\u001b[1m(\u001b[22m\u001b[90mmapexpr\u001b[39m::\u001b[0mtypeof(REPL.softscope), \u001b[90mmod\u001b[39m::\u001b[0mModule, \u001b[90mcode\u001b[39m::\u001b[0mString, \u001b[90mfilename\u001b[39m::\u001b[0mString\u001b[0m\u001b[1m)\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[90mBase\u001b[39m \u001b[90m.\\\u001b[39m\u001b[90m\u001b[4mloading.jl:2076\u001b[24m\u001b[39m\n",
      "    [7] \u001b[0m\u001b[1msoftscope_include_string\u001b[22m\u001b[0m\u001b[1m(\u001b[22m\u001b[90mm\u001b[39m::\u001b[0mModule, \u001b[90mcode\u001b[39m::\u001b[0mString, \u001b[90mfilename\u001b[39m::\u001b[0mString\u001b[0m\u001b[1m)\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[36mSoftGlobalScope\u001b[39m \u001b[90mC:\\Users\\mgnli\\.julia\\packages\\SoftGlobalScope\\u4UzH\\src\\\u001b[39m\u001b[90m\u001b[4mSoftGlobalScope.jl:65\u001b[24m\u001b[39m\n",
      "    [8] \u001b[0m\u001b[1mexecute_request\u001b[22m\u001b[0m\u001b[1m(\u001b[22m\u001b[90msocket\u001b[39m::\u001b[0mZMQ.Socket, \u001b[90mmsg\u001b[39m::\u001b[0mIJulia.Msg\u001b[0m\u001b[1m)\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[32mIJulia\u001b[39m \u001b[90mC:\\Users\\mgnli\\.julia\\packages\\IJulia\\Vo51o\\src\\\u001b[39m\u001b[90m\u001b[4mexecute_request.jl:67\u001b[24m\u001b[39m\n",
      "    [9] \u001b[0m\u001b[1m#invokelatest#2\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[90m.\\\u001b[39m\u001b[90m\u001b[4messentials.jl:892\u001b[24m\u001b[39m\u001b[90m [inlined]\u001b[39m\n",
      "   [10] \u001b[0m\u001b[1minvokelatest\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[90m.\\\u001b[39m\u001b[90m\u001b[4messentials.jl:889\u001b[24m\u001b[39m\u001b[90m [inlined]\u001b[39m\n",
      "   [11] \u001b[0m\u001b[1meventloop\u001b[22m\u001b[0m\u001b[1m(\u001b[22m\u001b[90msocket\u001b[39m::\u001b[0mZMQ.Socket\u001b[0m\u001b[1m)\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[32mIJulia\u001b[39m \u001b[90mC:\\Users\\mgnli\\.julia\\packages\\IJulia\\Vo51o\\src\\\u001b[39m\u001b[90m\u001b[4meventloop.jl:8\u001b[24m\u001b[39m\n",
      "   [12] \u001b[0m\u001b[1m(::IJulia.var\"#15#18\")\u001b[22m\u001b[0m\u001b[1m(\u001b[22m\u001b[0m\u001b[1m)\u001b[22m\n",
      "  \u001b[90m    @\u001b[39m \u001b[32mIJulia\u001b[39m \u001b[90mC:\\Users\\mgnli\\.julia\\packages\\IJulia\\Vo51o\\src\\\u001b[39m\u001b[90m\u001b[4meventloop.jl:38\u001b[24m\u001b[39m\n",
      "\u001b[0m\u001b[1mTest Summary: | \u001b[22m\u001b[91m\u001b[1mError  \u001b[22m\u001b[39m\u001b[36m\u001b[1mTotal  \u001b[22m\u001b[39m\u001b[0m\u001b[1mTime\u001b[22m\n",
      "Gene Tests    | \u001b[91m    1  \u001b[39m\u001b[36m    1  \u001b[39m\u001b[0m1.3s\n"
     ]
    },
    {
     "ename": "LoadError",
     "evalue": "\u001b[91mSome tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.\u001b[39m",
     "output_type": "error",
     "traceback": [
      "\u001b[91mSome tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.\u001b[39m",
      "",
      "Stacktrace:",
      " [1] finish(ts::Test.DefaultTestSet; print_results::Bool)",
      "   @ Test C:\\Users\\mgnli\\.julia\\juliaup\\julia-1.10.3+0.x64.w64.mingw32\\share\\julia\\stdlib\\v1.10\\Test\\src\\Test.jl:1195",
      " [2] finish(ts::Test.DefaultTestSet)",
      "   @ Test C:\\Users\\mgnli\\.julia\\juliaup\\julia-1.10.3+0.x64.w64.mingw32\\share\\julia\\stdlib\\v1.10\\Test\\src\\Test.jl:1170",
      " [3] macro expansion",
      "   @ C:\\Users\\mgnli\\.julia\\juliaup\\julia-1.10.3+0.x64.w64.mingw32\\share\\julia\\stdlib\\v1.10\\Test\\src\\Test.jl:1593 [inlined]",
      " [4] top-level scope",
      "   @ In[2]:23"
     ]
    }
   ],
   "source": [
    "## tests telomeres adn genes \n",
    "\n",
    "using Test \n",
    "\n",
    "function test_telomere_type()\n",
    "    a = Telomere()\n",
    "    @test typeof(a) == Telomere\n",
    "end\n",
    "\n",
    "function test_forward_reverse_gene()\n",
    "    @test isequal(Gene(0, \"d\", true), Gene(0, \"d\", false))\n",
    "end\n",
    "\n",
    "function test_telomere_repr()\n",
    "    @test show(Telomere()) == \".\"\n",
    "end\n",
    "\n",
    "function test_reverse_equality()\n",
    "    @test !isequal(Gene(0, \"D\", true), Gene(1, \"D\", true))\n",
    "end\n",
    "\n",
    "@testset \"Gene Tests\" begin\n",
    "    test_telomere_type()\n",
    "    test_forward_reverse_gene()\n",
    "    test_telomere_repr()\n",
    "    test_reverse_equality()\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Funcs \n",
    "\n",
    "\"\"\"\n",
    "checks for correct conditions \n",
    "    - only letters and dots\n",
    "    - even number of telomeres in both genomes (can't have the start of a chromosome without closing it)  \n",
    "\"\"\"\n",
    "function check_conditions(A::String, B::String)\n",
    "    set_A = Dict{Char, Int64}\n",
    "    set_B = Set{Char}()\n",
    "    tel_count_A = 0 \n",
    "    tel_count_B = 0 \n",
    "\n",
    "    for i in 1:length(A)\n",
    "        if A[i] == '.'\n",
    "            tel_count_A += 1\n",
    "        elseif isletter(A[i])\n",
    "            push!(set_A, lowercase(A[i])) # or uppercase, just keep consistent\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome A has non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    for i in 1:length(B) \n",
    "        if B[i] == '.'\n",
    "            tel_count_B += 1\n",
    "        elseif  isletter(A[i])\n",
    "            push!(set_B, lowercase(B[i]))\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome B has non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    if mod(tel_count_A, 2) != 0 || mod(tel_count_B, 2) != 0  \n",
    "         throw(ArgumentError(\"Error: Telomere error\"))\n",
    "        return false \n",
    "    elseif set_A != set_B\n",
    "        throw(ArgumentError(\"Error: Sets of genes in the two given genomes don't match\"))\n",
    "        return false\n",
    "    end \n",
    "\n",
    "\n",
    "    return true \n",
    "end \n",
    "\n",
    "# correct_conditions(\".abdc.\", \"abCd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: `AbstractGene` not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: `AbstractGene` not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ In[1]:1"
     ]
    }
   ],
   "source": [
    "function Chromosome(genes::Vector{AbstractGene}, id_set::Set{Int})\n",
    "    # print(\"entered Chromosome constructor for genes********************************************************\")\n",
    "    \n",
    "    for gene in genes\n",
    "        telomere = show(gene) == \".\"\n",
    "        if !telomere \n",
    "            if gene.id ∉ id_set \n",
    "                push!(id_set, gene.id)\n",
    "            else \n",
    "                throw(ArgumentError(\"There are duplicate genes with the same ID in this genome.\"))\n",
    "            end\n",
    "        end \n",
    "    end \n",
    "    content = genes\n",
    "\n",
    "    return Chromosome(content)\n",
    "\n",
    "end \n",
    "\n",
    "\n",
    "\n",
    "function Chromosome(::Any) \n",
    "    throw(TypeError(\"Chromosome must be a list of Gene instances with unique IDs or a string of genes.\"))\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### testing chromosome works \n",
    "\n",
    "# testing strings \n",
    "\n",
    "# id_counter = Ref{Int}(1)\n",
    "# id_to_char = Dict{Char, Int}()\n",
    "# char_to_id = Dict{Char, Vector{Int}}()\n",
    "\n",
    "# Chromosome(\".AbcA.\", id_counter, id_to_char, char_to_id)\n",
    "\n",
    "# print(\"\\n\\n******************\\nid_counter: \", id_counter[], \"\\n\", \"id_to_char\", id_to_char, \"\\n\",\"char_to_id: \", char_to_id)\n",
    "\n",
    "\n",
    "# testing genes \n",
    "x = Telomere()\n",
    "y = Gene(1, \"A\", true)\n",
    "z = Gene(2, \"B\", true)\n",
    "v = Telomere()\n",
    "\n",
    "id_set = Set{Int}() \n",
    "\n",
    "Chromosome([y, y, z], id_set)  ## why doesn't it find the right constructor? \n",
    "\n",
    "# testing rand \n",
    "# x = 1 \n",
    "# Chromosome(x) ## why doesn't it throw error??? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Gene head or tail end. Used in adjacencies\n",
    "\"\"\"\n",
    "@with_kw struct GeneEnd\n",
    "    gene::AbstractGene\n",
    "    head::Bool=true\n",
    "end\n",
    "\n",
    "\"\"\"Adjacency Data Structure\"\"\"\n",
    "mutable struct Adjacency \n",
    "    left_end_gene::GeneEnd\n",
    "    right_end_gene::GeneEnd\n",
    "    label::Vector{Gene}\n",
    "\n",
    "    Adjacency(left_end_gene::GeneEnd = nothing, right_end_gene::GeneEnd = nothing; label::Vector{Gene} = nothing) = new([left_end_gene, right_end_gene], left_end_gene, right_end_gene, label)\n",
    "end\n",
    "\n",
    "function Base.isequal(a::GeneEnd, b::GeneEnd)\n",
    "    return a.gene == b.gene && a.head == b.head\n",
    "end\n",
    "\n",
    "function Base.show(me::GeneEnd)\n",
    "    if me.head\n",
    "        print(string(me.gene))\n",
    "    else\n",
    "        print(string(me.gene) * \"*\")\n",
    "    end\n",
    "end\n",
    "\n",
    "function other_adjacency_end(me::GeneEnd, adj::Adjacency)\n",
    "    if adj.left_end_gene == me\n",
    "        return adj.right_end_gene\n",
    "    else\n",
    "        return adj.left_end_gene\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# AdjacencyGraph\n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    commonGenes::Set{AbstractGene}\n",
    "    adjA::Vector{Adjacency}\n",
    "    adjB::Vector{Adjacency}\n",
    "    cycles::Int\n",
    "    ab_paths::Int\n",
    "    a_runs::Int\n",
    "    b_runs::Int\n",
    "    run_potential::Int\n",
    "    indel_potential::Int\n",
    "end\n",
    "\n",
    "function AdjacencyGraph(A::Genome, B::Genome)\n",
    "    gene_set_a = Set(Iterators.flatten(A.data))\n",
    "    gene_set_b = Set(Iterators.flatten(B.data))\n",
    "    commonGenes = intersect(gene_set_a, gene_set_b)\n",
    "    if Telomere() in commonGenes\n",
    "        delete!(commonGenes, Telomere())\n",
    "    end\n",
    "    gene_set_a = setdiff(gene_set_a, commonGenes)\n",
    "    gene_set_b = setdiff(gene_set_b, commonGenes)\n",
    "\n",
    "    adjA = Adjacency[]\n",
    "    adjB = Adjacency[]\n",
    "    adj = [GeneEnd(Telomere()), GeneEnd(Telomere())]\n",
    "    adjacencies = [adjA, adjB]\n",
    "    reference_A = Dict{GeneEnd, Int}(GeneEnd(Telomere(), true) => 0)\n",
    "    reference_B = Dict{GeneEnd, Int}(GeneEnd(Telomere(), true) => 0)\n",
    "    references = [reference_A, reference_B]\n",
    "\n",
    "    for (i, genome) in enumerate([A, B])\n",
    "        adjacency_length = 0\n",
    "        for chromosome in genome.data\n",
    "            index = 0\n",
    "            adj_index = 0\n",
    "            current_gene = chromosome.data[index]\n",
    "            chromosome_genes = Set(chromosome.data)\n",
    "            adjacency_length += length(intersect(commonGenes, chromosome_Genes))\n",
    "            if chromosome.data[1] isa Telomere\n",
    "                adjacency_length += 1\n",
    "            end\n",
    "            while length(adjacencies[i]) < adjacency_length\n",
    "                label = AbstractGene[]\n",
    "                if current_gene isa Telomere\n",
    "                    adj[1] = GeneEnd(Telomere(), true)\n",
    "                else\n",
    "                    adj[1] = GeneEnd(current_gene, !current_gene.reverse)\n",
    "                end\n",
    "                next_gene = ifelse(index >= length(chromosome.data) - 1, chromosome.data[1], chromosome.data[index + 1])\n",
    "                while next_gene ∉ commonGene\n",
    "                    if next_gene isa Telomere\n",
    "                        break\n",
    "                    end\n",
    "                    push!(label, next_gene)\n",
    "                    index += 1\n",
    "                    next_gene = ifelse(index >= length(chromosome.data) - 1, chromosome.data[1], chromosome.data[index + 1])\n",
    "                end\n",
    "                if next_gene isa Telomere\n",
    "                    adj[2] = GeneEnd(Telomere(), true)\n",
    "                else\n",
    "                    adj[2] = GeneEnd(next_gene, next_gene.reverse)\n",
    "                end\n",
    "                push!(adjacencies[i], Adjacency(adj[1], adj[2], label))\n",
    "                if adj[1].gene != Telomere()\n",
    "                    references[i][adj[1]] = length(adjacencies[i])\n",
    "                end\n",
    "                if adj[2].gene != Telomere()\n",
    "                    references[i][adj[2]] = length(adjacencies[i])\n",
    "                end\n",
    "                current_gene = next_gene\n",
    "                index += 1\n",
    "                adj_index += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    to_visit_a_index = Set(1:length(adjA))\n",
    "    to_visit_b_index = Set(1:length(adjB))\n",
    "    visited_a_index = Set{Int}()\n",
    "    visited_b_index = Set{Int}()\n",
    "    cycles = 0\n",
    "    ab_paths = 0\n",
    "    a_runs = 0\n",
    "    on_a_run = false\n",
    "    b_runs = 0\n",
    "    on_b_run = false\n",
    "\n",
    "    while !isempty(to_visit_a_index)\n",
    "        current_adj_index = pop!(to_visit_a_index)\n",
    "        push!(visited_a_index, current_adj_index)\n",
    "        left_gene = adjA[current_adj_index].left_end_gene\n",
    "        right_gene = adjA[current_adj_index].right_end_gene\n",
    "        if !isempty(adjA[current_adj_index].label)\n",
    "            a_runs += 1\n",
    "            on_a_run = true\n",
    "        end\n",
    "        paths_end_on_a = [true, true]\n",
    "        for (i, current_gene) in enumerate([left_gene, right_gene])\n",
    "            a_side = true\n",
    "            next_adj_index = reference_B[current_gene]\n",
    "            if next_adj_index in visited_b_index\n",
    "                continue\n",
    "            end\n",
    "            while next_adj_index != nothing\n",
    "                current_adj_index = next_adj_index\n",
    "                a_side = !a_side\n",
    "                adj_side = a_side ? adjA : adjB\n",
    "                current_adj = adj_side[current_adj_index]\n",
    "                reference_side = a_side ? reference_B : reference_A\n",
    "                if a_side\n",
    "                    push!(visited_a_index, current_adj_index)\n",
    "                    delete!(to_visit_a_index, current_adj_index)\n",
    "                else\n",
    "                    push!(visited_b_index, current_adj_index)\n",
    "                    delete!(to_visit_b_index, current_adj_index)\n",
    "                end\n",
    "                if !isempty(current_adj.label)\n",
    "                    if !a_side && on_a_run\n",
    "                        on_b_run = true\n",
    "                        on_a_run = false\n",
    "                        b_runs += 1\n",
    "                    elseif !a_side && !on_b_run\n",
    "                        on_b_run = true\n",
    "                        b_runs += 1\n",
    "                    elseif a_side && on_b_run\n",
    "                        on_a_run = true\n",
    "                        on_b_run = false\n",
    "                        a_runs += 1\n",
    "                    elseif a_side && !on_a_run\n",
    "                        on_a_run = true\n",
    "                        a_runs += 1\n",
    "                    end\n",
    "                end\n",
    "                current_gene = current_gene == current_adj.left_end_gene ? current_adj.right_end_gene : current_adj.left_end_gene\n",
    "                next_adj_index = reference_side[current_gene]\n",
    "                if next_adj_index == nothing\n",
    "                    paths_end_on_a[i] = a_side\n",
    "                    break\n",
    "                elseif (a_side && next_adj_index in visited_b_index) || (!a_side && next_adj_index in visited_a_index)\n",
    "                    cycles += 1\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        if paths_end_on_a[1] != paths_end_on_a[2]\n",
    "            ab_paths += 1\n",
    "        end\n",
    "    end\n",
    "\n",
    "    run_potential = a_runs + b_runs\n",
    "    indel_potential = run_potential > 0 ? (run_potential + 1) ÷ 2 + ((run_potential ÷ 2) % 2) : 0\n",
    "\n",
    "    return AdjacencyGraph(commonGenes, adjA, adjB, cycles, ab_paths, a_runs, b_runs, run_potential, indel_potential)\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " \"\"\"The abstraction of a chromosome in DCJ model, a list of telomeres and genes\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genes: list of genes or string (req)\n",
    "        The object to be converted into a chromosome.\n",
    "\n",
    "    \"\"\"\n",
    "mutable struct Chromosome\n",
    "    genes::Vector{AbstractGene}\n",
    "    # gene_set::Set{String}\n",
    "end\n",
    "\n",
    "# mutable struct ChromosomeStatic\n",
    "#     string_d::Dict{String, Int}  # gene to ID \n",
    "#     uid_counter::Int\n",
    "#     ChromosomeStatic() = new(Dict{String, Int}(), 0)\n",
    "# end\n",
    "\n",
    "# const chromosome_static = ChromosomeStatic()\n",
    "\n",
    "function Chromosome(genes::Vector{})\n",
    "    # gene_set = Set{AbstractGene}()\n",
    "    content = AbstractGene[]\n",
    "    \n",
    "    # genes = list of genes \n",
    "    if isa(genes, Vector{AbstractGene})\n",
    "        content = genes\n",
    "     \n",
    "    # throw(TypeError(\"Chromosome must be a list of Gene instances or a string of unique genes.\"))\n",
    "\n",
    "    # parse string \n",
    "    # else \n",
    "    #     genes_str = collect(genes)\n",
    "    #     if !((genes_str[1] == '.') == (genes_str[end] == '.'))\n",
    "    #         throw(ArgumentError(\"Linear Chromosome must start and end with telomeres.\"))\n",
    "    #     end\n",
    "    #     for (i, s) in enumerate(genes_str)\n",
    "    #         dna = string(s)\n",
    "    #         rev = isuppercase(s)\n",
    "    #         telomere = s == '.'\n",
    "    #         # handle gene \n",
    "    #         if s in gene_set && !telomere\n",
    "    #             throw(ArgumentError(\"Duplicated genes are not allowed. ($s)\"))\n",
    "    #         elseif s in keys(chromosome_static.string_d)\n",
    "    #             uid = chromosome_static.string_d[s]\n",
    "    #             push!(gene_set, s)\n",
    "    #         else\n",
    "    #             chromosome_static.string_d[s] = chromosome_static.uid_counter\n",
    "    #             push!(gene_set, s)\n",
    "    #             uid = chromosome_static.uid_counter\n",
    "    #             chromosome_static.uid_counter += 1\n",
    "    #         end\n",
    "    #         # handle telomere\n",
    "    #         if telomere && !(i == 1 || i == length(genes_str))\n",
    "    #             throw(ArgumentError(\"Telomere cannot appear in middle of chromosome.\"))\n",
    "    #         elseif telomere\n",
    "    #             push!(content, Telomere())\n",
    "    #         else\n",
    "    #             push!(content, Gene(uid, dna, rev))\n",
    "    #         end\n",
    "    #     end\n",
    "    end\n",
    "    return Chromosome(content, gene_set)\n",
    "end\n",
    "\n",
    "function clear_chromosome_static()\n",
    "    chromosome_static.string_d = Dict{String, Int}()\n",
    "    chromosome_static.uid_counter = 0\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
