{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11809fe2",
   "metadata": {},
   "source": [
    "# Calculating DCJ distance between two unsigned, circular and/or linear genomes with no duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ea4c69",
   "metadata": {},
   "source": [
    "Credit:\n",
    "- https://github.com/mlliou112/py-dcj (python version)\n",
    "- ChatGPT for converting to Julia "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "221c64a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import statements\n",
    "using Base.Iterators\n",
    "using Parameters\n",
    "using DataStructures\n",
    "\n",
    "abstract type AbstractGene end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "91f5f99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# telomere\n",
    "\n",
    "@with_kw struct Telomere <: AbstractGene\n",
    "    reverse::Bool=false  # TODO\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Telomere, b)\n",
    "    return typeof(a) == typeof(b)\n",
    "end\n",
    "\n",
    "function Base.show(t::Telomere)\n",
    "    return \".\"\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b81dd52d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene\n",
    "\n",
    "@with_kw struct Gene <: AbstractGene\n",
    "    id::Int\n",
    "    dna::Char\n",
    "    reverse::Bool=false\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Gene, b::Gene)\n",
    "    return typeof(a) == typeof(b) && a.id == b.id\n",
    "end\n",
    "\n",
    "function Base.show(m::Gene)\n",
    "    if m.reverse\n",
    "        print(\"-\" * string(m.id))\n",
    "    else\n",
    "        print(string(m.id))\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7f20e9ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chromosome\n",
    "\n",
    "mutable struct Chromosome\n",
    "    genes::Vector{AbstractGene}\n",
    "end\n",
    "\n",
    "function Chromosome(genes::String, id_counter::Ref{Int64}, id_to_char::Dict{Int, Char}, char_to_id::Dict{Char, Int64}, target::Bool)\n",
    "    content = AbstractGene[]\n",
    "\n",
    "    if !((genes[1] == '.') == (genes[end] == '.'))\n",
    "        throw(ArgumentError(\"Linear geneosome must start and end with telomeres.\"))\n",
    "    end\n",
    "\n",
    "    for (i, gene) in pairs(genes)\n",
    "        dna = only(lowercase(string(gene)))\n",
    "        rev = isuppercase(gene)\n",
    "        telomere = gene == '.'        \n",
    "        if telomere \n",
    "            push!(content, Telomere())\n",
    "        else  # gene \n",
    "            id = id_counter[]\n",
    "\n",
    "            if target  \n",
    "                id_to_char[id] = gene\n",
    "                if gene in keys(char_to_id)\n",
    "                    push!(char_to_id[dna], id)\n",
    "                else \n",
    "                    char_to_id[dna] = id\n",
    "                end \n",
    "                id_counter[] += 1\n",
    "\n",
    "            else  # source str \n",
    "                ## assuming no duplicates\n",
    "                id = char_to_id[dna]\n",
    "            end\n",
    "            push!(content, Gene(id, dna, rev))\n",
    "        end \n",
    "    end    \n",
    "    \n",
    "    return Chromosome(content)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0db29f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genome\n",
    "\n",
    "mutable struct Genome\n",
    "    data::Vector{Chromosome}\n",
    "    id_to_char::Dict{Int, Char}\n",
    "    char_to_id::Dict{Char, Int}\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef463477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "string_to_genome (generic function with 1 method)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# funcs convert bt genome & str \n",
    "function genome_to_string(A)\n",
    "end\n",
    "\n",
    "\n",
    "function string_to_genome(s, id_counter, id_to_char, char_to_id, target)\n",
    "     \"\"\"\n",
    "    Converts string to type Genome \n",
    "    \"\"\"\n",
    "\n",
    "    chrom_list_str = Vector{String}()\n",
    "\n",
    "    chrom = \"\"\n",
    "    new_chrom = true \n",
    "    linear = false\n",
    "\n",
    "    for (i, g) in pairs(s)\n",
    "        # if creating a new chromosome \n",
    "        if new_chrom \n",
    "            if chrom == \"\"\n",
    "                linear = g == '.'\n",
    "            else  # chrom isn't empty \n",
    "                linear = chrom[1] == '.'\n",
    "            end \n",
    "\n",
    "            chrom = chrom * g\n",
    "            new_chrom = false  \n",
    "            if i == length(s) # last element is circular chrom \n",
    "                push!(chrom_list_str, chrom)\n",
    "               \n",
    "            end \n",
    "        \n",
    "        # appending more genes to chromosome \n",
    "        else  \n",
    "            if linear  \n",
    "                chrom = chrom * g \n",
    "                if g == '.' || i == length(s)  # find another telomere or end of str\n",
    "                    push!(chrom_list_str, chrom) \n",
    "                    chrom = \"\"\n",
    "                    new_chrom = true \n",
    "                end \n",
    "            else  # circular\n",
    "                if g =='.'  # close circular chrom (won't incude telomere in chrom)\n",
    "                    push!(chrom_list_str, chrom)\n",
    "                    chrom = \".\"  # put telomere in next chrom \n",
    "                    new_chrom = true \n",
    "                elseif i == length(s)  # reach end of str \n",
    "                    chrom = chrom * g\n",
    "                    push!(chrom_list_str, chrom)\n",
    "                    chrom = \".\"\n",
    "                    new_chrom = true \n",
    "                else \n",
    "                    chrom = chrom * g\n",
    "                end \n",
    "            end \n",
    "        end \n",
    "    end\n",
    "\n",
    "    # convert chrom_list_str (list of strings) to chrom_list\n",
    "    chrom_list = Chromosome[]\n",
    "    \n",
    "    for c_str in chrom_list_str\n",
    "        c = Chromosome(c_str, id_counter, id_to_char, char_to_id, target)\n",
    "        push!(chrom_list, c)\n",
    "    end \n",
    "    \n",
    "    return Genome(chrom_list, id_to_char, char_to_id)\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9b91ea5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene end (used in adj)\n",
    "\n",
    "@with_kw struct GeneEnd\n",
    "    gene::AbstractGene\n",
    "    head::Bool=true\n",
    "end\n",
    "\n",
    "# adjacency\n",
    "mutable struct Adjacency \n",
    "    left::GeneEnd\n",
    "    right::GeneEnd\n",
    "end\n",
    "\n",
    "function Base.isequal(a::GeneEnd, b::GeneEnd)\n",
    "    return a.id == b.id && a.head == b.head\n",
    "end\n",
    "\n",
    "function Base.show(ge::GeneEnd)\n",
    "    if ge.head\n",
    "        print(string(ge.gene.dna) *\":h\")\n",
    "    else\n",
    "        print(string(ge.gene.dna) * \":t\")\n",
    "    end\n",
    "\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Adjacency, b::Adjacency)\n",
    "     # order of the gene ends doesn't matter\n",
    "    if a.left == b.left && a.right == b.right || a.left == b.right && a.right == b.left  \n",
    "        return true \n",
    "    end \n",
    "    return false \n",
    "end \n",
    "\n",
    "function Base.show(adj::Adjacency)\n",
    "    left = \"\"\n",
    "    right = \"\"\n",
    "    \n",
    "    if adj.left.gene == Telomere() \n",
    "        left = \".\"\n",
    "    else \n",
    "        left_gene = adj.left\n",
    "        (left_gene.head == true) ? left = left_gene.gene.dna * \":h\" : left = left_gene.gene.dna * \":t\" \n",
    "    end\n",
    "    \n",
    "    if adj.right.gene == Telomere() \n",
    "        right = \".\"\n",
    "    else \n",
    "        right_gene = adj.right\n",
    "        (right_gene.head == true) ? right = right_gene.gene.dna * \":h\" : right = right_gene.gene.dna * \":t\"\n",
    "    end \n",
    "\n",
    "    print(\"(\", left, \",\", right, \")\")\n",
    "end \n",
    "\n",
    "function Base.show(adj::Adjacency, add_spacer::Bool)\n",
    "    left = \"\"\n",
    "    right = \"\"\n",
    "    \n",
    "    if adj.left.gene == Telomere() \n",
    "        left = \".\"\n",
    "    else \n",
    "        left_gene = adj.left\n",
    "        (left_gene.head == true) ? left = left_gene.gene.dna * \":h\" : left = left_gene.gene.dna * \":t\" \n",
    "    end\n",
    "    \n",
    "    if adj.right.gene == Telomere() \n",
    "        right = \".\"\n",
    "    else \n",
    "        right_gene = adj.right\n",
    "        (right_gene.head == true) ? right = right_gene.gene.dna * \":h\" : right = right_gene.gene.dna * \":t\"\n",
    "    end \n",
    "\n",
    "    if add_spacer \n",
    "        print(\"(\", left, \"---\", right, \")\")\n",
    "    else\n",
    "        print(\"(\", left, \",\", right, \")\")\n",
    "    end \n",
    "end \n",
    "\n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency})\n",
    "    for adj in adj_list \n",
    "        show(adj) \n",
    "    end \n",
    "end \n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency}, idx1::Int, idx2::Int)\n",
    "    r = min(idx1, idx2)\n",
    "    s = max(idx1, idx2)\n",
    "    for (i, adj) in pairs(adj_list)  \n",
    "        show(adj, i == r || i == s) \n",
    "    end \n",
    "end \n",
    "\n",
    "function Base.show(adj_set::Set{Adjacency})\n",
    "    for adj in adj_set \n",
    "        show(adj) \n",
    "    end \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "76b6ec27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_to_adj_listset (generic function with 1 method)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genome to adj list and set \n",
    "\n",
    "function genome_to_adj_listset_helper(chrom::Chromosome, genes::Vector{AbstractGene}, adj_list::Vector{Adjacency})\n",
    "    linear = genes[1] == Telomere()\n",
    "    \n",
    "    # handle first gene end \n",
    "    if linear \n",
    "        left_end_adj = GeneEnd(Telomere(), false)\n",
    "    else  # circular \n",
    "        left_end_adj = GeneEnd(genes[1], !genes[1].reverse)\n",
    "    end \n",
    "\n",
    "    # handle gene bt first and last \n",
    "    for gene in genes\n",
    "        if gene == Telomere() || gene == genes[1] \n",
    "            continue \n",
    "        end \n",
    "        reversed = gene.reverse  \n",
    "        \n",
    "        right_gene_adj = GeneEnd(gene, reversed)  # if reversed, right_end_adj is head (& left_end_adj is tail) \n",
    "        adj = Adjacency(left_end_adj, right_gene_adj)\n",
    "\n",
    "        push!(adj_list, adj)\n",
    "\n",
    "        left_end_adj = GeneEnd(gene, !reversed)\n",
    "    end \n",
    "\n",
    "    # handle last gene end \n",
    "    if linear \n",
    "        right_end_adj = GeneEnd(Telomere(), false)\n",
    "    else   \n",
    "        right_end_adj = GeneEnd(genes[1], genes[1].reverse)\n",
    "    end\n",
    "\n",
    "    return left_end_adj, right_end_adj\n",
    "\n",
    "end \n",
    "\n",
    "\n",
    "function genome_to_adj_listset(genome::Genome)\n",
    "    adj_list = Vector{Adjacency}()\n",
    "    adj_set = Set{Adjacency}()\n",
    "\n",
    "    for chrom in genome.data\n",
    "        genes = chrom.genes\n",
    "        reversed = genes[1].reverse  \n",
    "        \n",
    "        if length(genes) == 1  # single gene edge case   \n",
    "            left_end_adj = GeneEnd(genes[1], reversed)  # if reversed, left_end_adj is head (& right_end_adj is tail) \n",
    "            right_end_adj = GeneEnd(genes[1], !reversed) \n",
    "        elseif length(genes) == 2  # two gene edge case \n",
    "            gene1 = genes[1]\n",
    "            gene2 = genes[2]\n",
    "\n",
    "            left_end_adj = GeneEnd(gene1, !gene1.reverse)\n",
    "            right_end_adj = GeneEnd(gene2, gene2.reverse) \n",
    "            adj = Adjacency(left_end_adj, right_end_adj)\n",
    "            push!(adj_list, adj)\n",
    "\n",
    "            reversed = genes[2].reverse \n",
    "            left_end_adj = GeneEnd(gene2, !gene2.reverse)\n",
    "            right_end_adj = GeneEnd(gene1, gene1.reverse) \n",
    "        else # > 2 gene in chrom  \n",
    "            left_end_adj, right_end_adj = genome_to_adj_listset_helper(chrom, genes, adj_list)\n",
    "        end \n",
    "        \n",
    "        adj = Adjacency(left_end_adj, right_end_adj)\n",
    "        push!(adj_list, adj)\n",
    "        push!(adj_set, adj)\n",
    "    end \n",
    "    return adj_list, adj_set\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d19634ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "process_adj_list (generic function with 1 method)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process adj list\n",
    "\n",
    "function process_adj_list_helper(ge::GeneEnd, idx::Ref{Int}, gid_to_loc::DefaultDict{Int, Vector{Int}})\n",
    "    if ge.gene == Telomere() \n",
    "        return\n",
    "    end  \n",
    "\n",
    "    if ge.head == true\n",
    "        gid_to_loc[ge.gene.id][2] = idx[]\n",
    "    else \n",
    "        gid_to_loc[ge.gene.id][1] = idx[]\n",
    "    end\n",
    "end \n",
    "\n",
    "function process_adj_list(adj_list:: Vector{Adjacency})\n",
    "    geneid_to_location = DefaultDict{Int, Vector{Int}}(() -> zeros(Int, 2)) # tail = idx 1, head = 2 in array\n",
    "    idx = Ref{Int}(1)\n",
    "\n",
    "    for adj in adj_list\n",
    "        process_adj_list_helper(adj.left, idx, geneid_to_location)\n",
    "        process_adj_list_helper(adj.right, idx, geneid_to_location)\n",
    "        idx[] += 1\n",
    "    end \n",
    "    \n",
    "    return geneid_to_location \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f5103360",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "reassign_locs (generic function with 1 method)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# helpers for dcj operations and distance\n",
    "\n",
    "# finds target gene end in source adjacency list \n",
    "# returns index and left/right in the source adj list\n",
    "function find_tar_ge_in_src_adjlist(target_ge::GeneEnd, src_gid_to_l::DefaultDict{Int, Vector{Int}}, tar_gid_to_l::DefaultDict{Int, Vector{Int}}, src_adj_list::Vector{Adjacency})\n",
    "    (target_ge.head == true) ? th_idx = 2 : th_idx = 1\n",
    "    # (left_tar_ge == true) ? tar_lr = 'L' : tar_lr = 'R'\n",
    "        \n",
    "    gene_id = target_ge.gene.id\n",
    "\n",
    "    src_ge_idx = src_gid_to_l[gene_id][th_idx]\n",
    "\n",
    "    adj = src_adj_list[src_ge_idx]\n",
    "    if adj.left.gene.id == gene_id\n",
    "        src_ge = adj.left\n",
    "        # src_lr = 'L'\n",
    "    else \n",
    "        src_ge = adj.right \n",
    "        # src_lr = 'R' \n",
    "    end \n",
    "\n",
    "    # print(\"\\n   GENEEND: \")\n",
    "    # show(target_ge)\n",
    "    # print(\"\\n     src genome idx: \", src_ge_idx, \" \", src_lr)\n",
    "    # print(\"\\n     target genome idx: \", tar_gid_to_l[gene_id][th_idx], \" \", tar_lr, \"\\n\")\n",
    "\n",
    "    return [src_ge, src_ge_idx]\n",
    "end \n",
    "\n",
    "function other_adjacency_end(ge::GeneEnd, adj::Adjacency)\n",
    "    if adj.left == ge\n",
    "        return adj.right\n",
    "    else\n",
    "        return adj.left\n",
    "    end \n",
    "end\n",
    "\n",
    "function combine_ge(u::GeneEnd, u_idx::Int, v::GeneEnd, v_idx::Int)\n",
    "    if u_idx < v_idx \n",
    "        u_lt_v = true \n",
    "    else \n",
    "        u_lt_v = false \n",
    "    end \n",
    "\n",
    "    if !u_lt_v \n",
    "        return Adjacency(v, u)\n",
    "    else \n",
    "        return Adjacency(u, v)\n",
    "    end\n",
    "\n",
    "end \n",
    "\n",
    "function reassign_locs(gid_to_l, adj_list, u_idx, v_idx)\n",
    "    for (i, adj) in pairs(view(adj_list, min(u_idx, v_idx):max(u_idx, v_idx)))\n",
    "        left = adj.left \n",
    "        right = adj.right \n",
    "\n",
    "        idx = Ref{Int}(i + min(u_idx, v_idx) - 1)\n",
    "\n",
    "        process_adj_list_helper(left, idx, gid_to_l)\n",
    "        process_adj_list_helper(right, idx, gid_to_l)\n",
    "    end \n",
    "end \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4d3a3511",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRC ADJ LIST\n",
      "(a:h,b:t)(b:h,c:t)(c:h,a:t)\n",
      "|\n",
      "v\n",
      "TARGET ADJ LIST\n",
      "(c:h,b:t)(b:h,a:t)(a:h,c:t)\n",
      "\n",
      "*************\n",
      "indexing through target adj list...\n",
      "\n",
      "(a:h,b:t)(b:h,c:t)(c:h,a:t)\n",
      "ADJ 1 :::: (c:h,b:t)\n",
      "\n",
      "(a:h,b:t)(b:h,c:t)(c:h,a:t)\n",
      "ADJ 2 :::: (b:h,a:t)\n",
      "\n",
      "(a:h,b:t)(b:h,c:t)(c:h,a:t)\n",
      "ADJ 3 :::: (a:h,c:t)"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function update_adj_set(p::GeneEnd, q::GeneEnd, u::GeneEnd, v::GeneEnd, u_idx::Int, v_idx::Int, src_adj_list::Vector{Adjacency}, src_adj_set::Set{Adjacency}) \n",
    "    #  replace adj u and v in A by ( {p, q} and (u\\{p}) U (v\\{q}) )\n",
    "    pq = combine_ge(p, u_idx, q, v_idx)\n",
    "\n",
    "    other_ge_u = other_adjacency_end(u, src_adj_list[u_idx])\n",
    "    other_ge_v = other_adjacency_end(v, src_adj_list[v_idx])\n",
    "    excluding_pq = combine_ge(other_ge_u, u_idx, other_ge_v, v_idx) \n",
    "    \n",
    "    #  remove adj with u, v in src_adj_set \n",
    "    delete!(src_adj_set, src_adj_list[u_idx])\n",
    "    delete!(src_adj_set, src_adj_list[v_idx])\n",
    "    \n",
    "    #  add adjusted adjacencies to src_adj_set\n",
    "    push!(src_adj_set, pq)\n",
    "    push!(src_adj_set, excluding_pq)\n",
    "\n",
    "    # print(\"  src_adj_set:\")\n",
    "    # show(src_adj_set)\n",
    "    # print(\"\\n\")\n",
    "end \n",
    "\n",
    "function update_adj_list_gid_loc_dict() \n",
    "    # update adj lists \n",
    "        print(\"\\nDCJ Operation (Cycle) :: \")\n",
    "        show(src_adj_list, u_idx, v_idx)\n",
    "        \n",
    "        print(\"\\n\")\n",
    "        show(pq)\n",
    "        show(excluding_pq)\n",
    "        if u_idx < v_idx \n",
    "            src_adj_list[u_idx] = pq\n",
    "            src_adj_list[v_idx] = excluding_pq\n",
    "            reverse!(view(src_adj_list, u_idx+1:v_idx-1))\n",
    "        else \n",
    "            src_adj_list[u_idx] = excluding_pq\n",
    "            src_adj_list[v_idx] = pq\n",
    "            reverse!(view(src_adj_list, v_idx+1:u_idx-1))\n",
    "        end  \n",
    "    \n",
    "        # update gid-->loc dicts \n",
    "        reassign_locs(src_gid_to_l, src_adj_list, u_idx, v_idx)\n",
    "end \n",
    "\n",
    "# find dcj operations and distance\n",
    "\n",
    "function find_dcj_dist_ops(src_adj_list::Vector{Adjacency}, tar_adj_list::Vector{Adjacency}, src_gid_to_l::DefaultDict{Int, Vector{Int}}, tar_gid_to_l::DefaultDict{Int, Vector{Int}}, src_adj_set::Set{Adjacency}, target_adj_set::Set{Adjacency})\n",
    "    count = 0  \n",
    "    # cycles = 0 \n",
    "    # odd_paths = 0 \n",
    "    telomere_idxs = Vector{Int}\n",
    "\n",
    "    print(\"indexing through target adj list...\")\n",
    "    \n",
    "    # for each adj {p, q} in target genome \n",
    "    for (i, adj) in pairs(tar_adj_list) \n",
    "        print(\"\\n\\n\")\n",
    "        show(src_adj_list)\n",
    "         \n",
    "        print(\"\\nADJ \", i, \" :::: \")\n",
    "        show(adj)\n",
    "\n",
    "        p = adj.left  \n",
    "        q = adj.right  \n",
    "\n",
    "        if p.gene == Telomere() || q.gene == Telomere()  # telomeres handled in next loop\n",
    "            push!(telomere_idxs, i)  \n",
    "        end \n",
    "\n",
    "        #  let u be element of genome A that contains p \n",
    "        #  let v be element of genome A that contains q \n",
    "        u, u_idx = find_tar_ge_in_src_adjlist(p, src_gid_to_l, tar_gid_to_l, src_adj_list)  \n",
    "        v, v_idx = find_tar_ge_in_src_adjlist(q, src_gid_to_l, tar_gid_to_l, src_adj_list)  \n",
    "        \n",
    "        #  if u != v, replace u and v in A by {p, q} and (u\\{p}) U (v\\{q})\n",
    "        if u_idx != v_idx             \n",
    "            update_adj_set(p, q, u, v, u_idx, v_idx, src_adj_list, src_adj_set)\n",
    "            # update_adj_list_gid_loc_dict()\n",
    "            \n",
    "            count += 1 \n",
    "        end\n",
    "        # cycles += 1  # TODO cycle counting     \n",
    "    end \n",
    "\n",
    "    # for each telomere {p} in genome B \n",
    "    for idx in telomere_idxs\n",
    "        telo_adj = tar_adj_list[idx]\n",
    "        telo_ge = other_adjacency_end(telo_adj, Telomere())\n",
    "\n",
    "        #  let u be element of genome A that contains p \n",
    "        u, u_idx = find_tar_ge_in_src_adjlist(p, src_gid_to_l, tar_gid_to_l, src_adj_list)  \n",
    "\n",
    "        #  if u is an adjacency, then replace u in A by {p} and (u\\{p})\n",
    "        if other_adjacency_end(tar_adj_list[u_idx], u) != Telomere() \n",
    "            # update_adj_set(p, u, u_idx,src_adj_list, src_adj_set)\n",
    "            # update_adj_list_gid_loc_dict()\n",
    "\n",
    "            count += 1 \n",
    "        end \n",
    "\n",
    "        #  odd paths++  # TODO odd path counting \n",
    "    end \n",
    "\n",
    "    return count \n",
    "    # return cycles, odd_paths\n",
    "end \n",
    "\n",
    "\n",
    "\n",
    "src = \"abc\"\n",
    "target = \"cba\"\n",
    "\n",
    "calculate_distance(src, target)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "08982ad1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "calculate_distance (generic function with 1 method)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adjacency graph \n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    cycles::Int\n",
    "    odd_paths::Int\n",
    "end\n",
    "\n",
    "function Base.show(dict::DefaultDict{Int, Vector{Int}}, id_to_char::Dict{Int, Char})\n",
    "    for (key, value) in dict\n",
    "        # println(\"$(key.dna) (id:$(key.id)) => $(value)\")\n",
    "        println(id_to_char[key], \" $(key) => $(value)\")\n",
    "    end\n",
    "end \n",
    "\n",
    "function AdjacencyGraph(src::Genome, target::Genome)\n",
    "\n",
    "    # process genomes (not explicitly creating the adj graph)    \n",
    "    src_adj_list, src_adj_set = genome_to_adj_listset(src)        \n",
    "    target_adj_list, target_adj_set = genome_to_adj_listset(target)        \n",
    "    \n",
    "    print(\"SRC ADJ LIST\\n\")\n",
    "    show(src_adj_list)\n",
    "    print(\"\\n|\\nv\\nTARGET ADJ LIST\\n\")\n",
    "    show(target_adj_list)\n",
    "    print(\"\\n\\n*************\\n\")\n",
    "    \n",
    "    src_geneid_to_location = process_adj_list(src_adj_list)   \n",
    "    target_geneid_to_location = process_adj_list(target_adj_list)\n",
    "    \n",
    "    if src.id_to_char != target.id_to_char\n",
    "        throw(ArgumentError(\"ERROR: Source ID to char dict is not equal to target's.\"))\n",
    "    end \n",
    "\n",
    "    # print(\"SRC GENE --> LOCATION DICT\\n\")\n",
    "    # show(src_geneid_to_location, src.id_to_char)  \n",
    "    # print(\"\\n\\n\", \"TARGET GENE --> LOCATION DICT\\n\")\n",
    "    # show(target_geneid_to_location, src.id_to_char)\n",
    "    # print(\"\\n*************\\n\\n\")\n",
    "\n",
    "    # use adj graph representation to find dcj distance and operations \n",
    "    # cycles, odd_paths = \n",
    "    find_dcj_dist_ops(src_adj_list, target_adj_list, src_geneid_to_location, target_geneid_to_location, src_adj_set, target_adj_set) \n",
    "    \n",
    "    # return AdjacencyGraph(cycles, odd_paths)\n",
    "end \n",
    "\n",
    "\"\"\"\n",
    "assumptions \n",
    "genomes A and B have the same letters of the same multiplicity; telomeres don't matter \n",
    "no empty chromosomes (2 telomeres ..)\n",
    "\"\"\"\n",
    "\n",
    "function calculate_distance(src::String, target::String)\n",
    "    # check_conditions(src, target)\n",
    "\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}()\n",
    "\n",
    "    target_genome = string_to_genome(target, id_counter, id_to_char, char_to_id, true)\n",
    "    src_genome = string_to_genome(src, id_counter, id_to_char, char_to_id, false)\n",
    "\n",
    "    ag = AdjacencyGraph(src_genome, target_genome)\n",
    "\n",
    "    # dcj_distance = src_genome.num_genes - ag.cycles - (ag.ab_paths / 2)  # or target_genome.num_genes\n",
    "\n",
    "    # return dcj_distance\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
