{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11809fe2",
   "metadata": {},
   "source": [
    "# Calculating DCJ distance between two unsigned, circular and/or linear genomes with no duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ea4c69",
   "metadata": {},
   "source": [
    "Credit:\n",
    "- https://github.com/mlliou112/py-dcj (python version)\n",
    "- ChatGPT for converting to Julia "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "221c64a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Base.Iterators\n",
    "using Parameters\n",
    "using DataStructures\n",
    "\n",
    "abstract type AbstractGene end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "91f5f99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Representation for a telomere in DCJ model\n",
    "\"\"\"\n",
    "\n",
    "@with_kw struct Telomere <: AbstractGene\n",
    "    reverse::Bool=false  # TODO\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Telomere, b)\n",
    "    return typeof(a) == typeof(b)\n",
    "end\n",
    "\n",
    "function Base.show(t::Telomere)\n",
    "    return \".\"\n",
    "end\n",
    "\n",
    "# x = Telomere()\n",
    "# show(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b81dd52d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Representation for DNA fragment (a gene) in DCJ model.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    id: int (required)\n",
    "        unique id\n",
    "\n",
    "    dna: str\n",
    "        What dna fragment does this gene represent?\n",
    "\n",
    "    reverse: boolean\n",
    "        Is the DNA fragment reversed?\n",
    "\"\"\"\n",
    "@with_kw struct Gene <: AbstractGene\n",
    "    id::Int\n",
    "    dna::String=\"\"\n",
    "    reverse::Bool=false\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Gene, b::Gene)\n",
    "    return typeof(a) == typeof(b) && a.id == b.id\n",
    "end\n",
    "\n",
    "function Base.show(m::Gene)\n",
    "    if m.reverse\n",
    "        print(\"-\" * string(m.id))\n",
    "    else\n",
    "        print(string(m.ud))\n",
    "    end\n",
    "end\n",
    "\n",
    "# y = Gene(1, \"d\", true)\n",
    "# show(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7f20e9ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " \"\"\"The abstraction of a chromosome in DCJ model, a list of telomeres and genes\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genes: list of genes or string (req)\n",
    "        The object to be converted into a circular or linear chromosome.\n",
    "\n",
    "    \"\"\"\n",
    "mutable struct Chromosome\n",
    "    genes::Vector{AbstractGene}\n",
    "end\n",
    "\n",
    "function Chromosome(genes::String, id_counter::Ref{Int64}, id_to_char::Dict{Int, Char}, char_to_id::Dict{Char, Int64}, target::Bool)\n",
    "    content = AbstractGene[]\n",
    "\n",
    "    if !((genes[1] == '.') == (genes[end] == '.'))\n",
    "        throw(ArgumentError(\"Linear geneosome must start and end with telomeres.\"))\n",
    "    end\n",
    "\n",
    "    for (i, gene) in pairs(genes)\n",
    "        dna = string(gene)\n",
    "        rev = isuppercase(gene)\n",
    "        telomere = gene == '.'        \n",
    "        if telomere \n",
    "            push!(content, Telomere())\n",
    "        else  # gene \n",
    "            id = id_counter[]\n",
    "\n",
    "            if target  \n",
    "                id_to_char[id] = gene\n",
    "                if gene in keys(char_to_id)\n",
    "                    push!(char_to_id[lowercase(gene)], id)\n",
    "                else \n",
    "                    char_to_id[lowercase(gene)] = id\n",
    "                end \n",
    "                id_counter[] += 1\n",
    "\n",
    "            else  # source str \n",
    "                ## assuming no duplicates\n",
    "                id = char_to_id[lowercase(gene)]\n",
    "            end\n",
    "            push!(content, Gene(id, dna, rev))\n",
    "        end \n",
    "    end    \n",
    "    \n",
    "    return Chromosome(content)\n",
    "end\n",
    "\n",
    "\n",
    "# x = \"b\"\n",
    "# id_counter = Ref{Int}(1)\n",
    "\n",
    "# Chromosome(x, id_counter, Dict{Int, Char}(), Dict{Char, Int64}(), true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0db29f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genome\n",
    "mutable struct Genome\n",
    "    data::Vector{Chromosome}\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef463477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "string_to_genome (generic function with 1 method)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function genome_to_string(A)\n",
    "end\n",
    "\n",
    "\n",
    "function string_to_genome(s, id_counter, id_to_char, char_to_id, target)\n",
    "     \"\"\"\n",
    "    Converts string to type Genome \n",
    "    \"\"\"\n",
    "\n",
    "    chrom_list_str = Vector{String}()\n",
    "\n",
    "    chrom = \"\"\n",
    "    new_chrom = true \n",
    "    linear = false\n",
    "\n",
    "    for (i, g) in pairs(s)\n",
    "        # if creating a new chromosome \n",
    "        if new_chrom \n",
    "            if chrom == \"\"\n",
    "                linear = g == '.'\n",
    "            else  # chrom isn't empty \n",
    "                linear = chrom[1] == '.'\n",
    "            end \n",
    "\n",
    "            chrom = chrom * g\n",
    "            new_chrom = false  \n",
    "            if i == length(s) # last element is circular chrom \n",
    "                push!(chrom_list_str, chrom)\n",
    "               \n",
    "            end \n",
    "        \n",
    "        # appending more genes to chromosome \n",
    "        else  \n",
    "            if linear  \n",
    "                chrom = chrom * g \n",
    "                if g == '.' || i == length(s)  # find another telomere or end of str\n",
    "                    push!(chrom_list_str, chrom) \n",
    "                    chrom = \"\"\n",
    "                    new_chrom = true \n",
    "                end \n",
    "            else  # circular\n",
    "                if g =='.'  # close circular chrom (won't incude telomere in chrom)\n",
    "                    push!(chrom_list_str, chrom)\n",
    "                    chrom = \".\"  # put telomere in next chrom \n",
    "                    new_chrom = true \n",
    "                elseif i == length(s)  # reach end of str \n",
    "                    chrom = chrom * g\n",
    "                    push!(chrom_list_str, chrom)\n",
    "                    chrom = \".\"\n",
    "                    new_chrom = true \n",
    "                else \n",
    "                    chrom = chrom * g\n",
    "                end \n",
    "            end \n",
    "        end \n",
    "    end\n",
    "\n",
    "    # convert chrom_list_str (list of strings) to chrom_list\n",
    "    chrom_list = Chromosome[]\n",
    "    \n",
    "    for c_str in chrom_list_str\n",
    "        c = Chromosome(c_str, id_counter, id_to_char, char_to_id, target)\n",
    "        push!(chrom_list, c)\n",
    "    end \n",
    "\n",
    "    return Genome(chrom_list)\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9b91ea5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Gene head or tail end. Used in adjacencies\n",
    "\"\"\"\n",
    "@with_kw struct GeneEnd\n",
    "    gene::AbstractGene\n",
    "    head::Bool=true\n",
    "end\n",
    "\n",
    "\"\"\"Adjacency Data Structure\"\"\"\n",
    "mutable struct Adjacency \n",
    "    left::GeneEnd\n",
    "    right::GeneEnd\n",
    "end\n",
    "\n",
    "function Base.isequal(a::GeneEnd, b::GeneEnd)\n",
    "    return a.gene == b.gene && a.head == b.head\n",
    "end\n",
    "\n",
    "function Base.show(ge::GeneEnd)\n",
    "    if ge.head\n",
    "        print(string(ge.gene))\n",
    "    else\n",
    "        print(string(ge.gene) * \"*\")\n",
    "    end\n",
    "end\n",
    "\n",
    "function Base.show(adj::Adjacency)\n",
    "    left = \"\"\n",
    "    right = \"\"\n",
    "    \n",
    "    if adj.left.gene == Telomere() \n",
    "        left = \".\"\n",
    "    else \n",
    "        left_gene = adj.left\n",
    "        (left_gene.head == true) ? left = left_gene.gene.dna * \":h\" : left = left_gene.gene.dna * \":t\" \n",
    "    end\n",
    "    \n",
    "    if adj.right.gene == Telomere() \n",
    "        right = \".\"\n",
    "    else \n",
    "        right_gene = adj.right\n",
    "        (right_gene.head == true) ? right = right_gene.gene.dna * \":h\" : right = right_gene.gene.dna * \":t\"\n",
    "    end \n",
    "\n",
    "    print(\"(\", left, \",\", right, \")\")\n",
    "end \n",
    "\n",
    "function Base.show(adj_list::Vector{Adjacency})\n",
    "    for adj in adj_list \n",
    "        show(adj) \n",
    "    end \n",
    "end \n",
    "\n",
    "# function other_adjacency_end(me::GeneEnd, adj::Adjacency)\n",
    "#     if adj.left == me\n",
    "#         return adj.right\n",
    "#     else\n",
    "#         return adj.left\n",
    "#     end\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "76b6ec27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_to_adj_list (generic function with 1 method)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function genome_to_adj_list_helper(chrom::Chromosome, genes::Vector{AbstractGene}, reversed::Bool, adj_list::Vector{Adjacency})\n",
    "    linear = genes[1] == Telomere()\n",
    "    \n",
    "    # handle first gene end \n",
    "    if linear \n",
    "        left_end_adj = GeneEnd(Telomere(), false)\n",
    "    else  # circular \n",
    "        left_end_adj = GeneEnd(genes[1], !reversed)\n",
    "    end \n",
    "\n",
    "    # handle chrom bt first and last \n",
    "    for gene in genes\n",
    "        if gene == Telomere() || gene == genes[1] \n",
    "            continue \n",
    "        end \n",
    "        reversed = gene.reverse  \n",
    "        \n",
    "        right_gene_adj = GeneEnd(gene, reversed)  # if reversed, right_end_adj is head (& left_end_adj is tail) \n",
    "        adj = Adjacency(left_end_adj, right_gene_adj)\n",
    "\n",
    "        push!(adj_list, adj)\n",
    "\n",
    "        left_end_adj = GeneEnd(gene, !reversed)\n",
    "    end \n",
    "\n",
    "    # handle last gene end \n",
    "    if linear \n",
    "        right_end_adj = GeneEnd(Telomere(), false)\n",
    "\n",
    "    else   \n",
    "        right_end_adj = GeneEnd(genes[1], reversed)\n",
    "    end\n",
    "\n",
    "    return left_end_adj, right_end_adj\n",
    "\n",
    "end \n",
    "\n",
    "\n",
    "function genome_to_adj_list(genome::Genome)\n",
    "    adj_list = Vector{Adjacency}()\n",
    "\n",
    "    for chrom in genome.data\n",
    "        genes = chrom.genes\n",
    "        reversed = genes[1].reverse  \n",
    "        \n",
    "        if length(genes) == 1  # single gene edge case   \n",
    "            left_end_adj = GeneEnd(genes[1], reversed)  # if reversed, left_end_adj is head (& right_end_adj is tail) \n",
    "            right_end_adj = GeneEnd(genes[1], !reversed) \n",
    "        else # > 1 gene in chrom  \n",
    "            left_end_adj, right_end_adj = genome_to_adj_list_helper(chrom, genes, reversed, adj_list)\n",
    "        end \n",
    "        adj = Adjacency(left_end_adj, right_end_adj)\n",
    "        push!(adj_list, adj)\n",
    "    end \n",
    "    return adj_list\n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d19634ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "process_adj_list (generic function with 1 method)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function process_adj_list_helper(ge::GeneEnd, idx::Ref{Int}, gene_to_location::DefaultDict{Gene, Vector{Int}})\n",
    "    if ge.gene == Telomere() \n",
    "        return\n",
    "    end  \n",
    "\n",
    "    if ge.head == true\n",
    "        gene_to_location[ge.gene][2] = idx[]\n",
    "    else \n",
    "        gene_to_location[ge.gene][1] = idx[]\n",
    "    end\n",
    "end \n",
    "\n",
    "function process_adj_list(adj_list:: Vector{Adjacency})\n",
    "    gene_to_location = DefaultDict{Gene, Vector{Int}}(() -> zeros(Int, 2)) # tail = idx 1, head = 2 in array\n",
    "    idx = Ref{Int}(1)\n",
    "\n",
    "    for adj in adj_list\n",
    "        process_adj_list_helper(adj.left, idx, gene_to_location)\n",
    "        process_adj_list_helper(adj.right, idx, gene_to_location)\n",
    "        idx[] += 1\n",
    "    end \n",
    "    \n",
    "    return gene_to_location \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "08982ad1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DefaultDict{Gene, Vector{Int64}, var\"#12#13\"} with 5 entries:\n",
       "  Gene(5, \"e\", false) => [4, 5]\n",
       "  Gene(3, \"c\", false) => [2, 3]\n",
       "  Gene(4, \"d\", false) => [3, 4]\n",
       "  Gene(1, \"a\", false) => [5, 1]\n",
       "  Gene(2, \"b\", false) => [1, 2]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutable struct AdjacencyGraph\n",
    "    src_adj_list::Vector{Adjacency}\n",
    "    target_adj_list::Vector{Adjacency}\n",
    "    \n",
    "    src_gene_to_location::DefaultDict{Gene, Vector{Int}}\n",
    "    target_gene_to_locationJ::DefaultDict{Gene, Vector{Int}}\n",
    "\n",
    "    # cycles::Int\n",
    "    # odd_paths::Int\n",
    "    \n",
    "    # commonGenes::Set{AbstractGene}\n",
    "    # ab_paths::Int\n",
    "    # a_runs::Int\n",
    "    # b_runs::Int\n",
    "    # run_potential::Int\n",
    "end\n",
    "\n",
    "function AdjacencyGraph(src::Genome, target::Genome)\n",
    "\n",
    "    # process genomes (not explicitly creating the adj graph)    \n",
    "    src_adj_list = genome_to_adj_list(src)        \n",
    "    target_adj_list = genome_to_adj_list(target)        \n",
    "  \n",
    "    # show(src_adj_list)\n",
    "    # print(\"\\n\")\n",
    "    # show(target_adj_list)\n",
    "    \n",
    "    src_gene_to_location = process_adj_list(src_adj_list)\n",
    "    target_gene_to_location = process_adj_list(target_adj_list)\n",
    "    \n",
    "\n",
    "    \n",
    "    # return AdjacencyGraph(src_adj_list, target_adj_list, src_gene_to_location, target_gene_to_location)\n",
    "end \n",
    "\n",
    "\"\"\"\n",
    "genomes A and B have the same letters of the same multiplicity; telomeres don't matter \n",
    "no empty chromosomes (2 telomeres ..)\n",
    "\"\"\"\n",
    "\n",
    "function calculate_distance(src::String, target::String)\n",
    "    # check_conditions(src, target)\n",
    "\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}()\n",
    "\n",
    "    target_genome = string_to_genome(target, id_counter, id_to_char, char_to_id, true)\n",
    "    src_genome = string_to_genome(src, id_counter, id_to_char, char_to_id, false)\n",
    "\n",
    "    ag = AdjacencyGraph(src_genome, target_genome)\n",
    "\n",
    "    # dcj_distance = length(ag.commonGenes) - ag.cycles - ag.ab_paths / 2\n",
    "    # operations = TODO \n",
    "end\n",
    "\n",
    "src = \".ba.cde\"\n",
    "target = \"abcde\"\n",
    "calculate_distance(src, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa3fe2e4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
