{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11809fe2",
   "metadata": {},
   "source": [
    "# Calculating DCJ distance between two unsigned, circular and/or linear genomes with no duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ea4c69",
   "metadata": {},
   "source": [
    "Credit:\n",
    "- https://github.com/mlliou112/py-dcj (python version)\n",
    "- ChatGPT for converting to Julia "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "221c64a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Base.Iterators\n",
    "using Parameters\n",
    "\n",
    "abstract type AbstractMarker end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "91f5f99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Representation for a telomere in DCJ model\n",
    "\"\"\"\n",
    "\n",
    "@with_kw struct Telomere <: AbstractMarker\n",
    "    reverse::Bool=false\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Telomere, b)\n",
    "    return typeof(a) == typeof(b)\n",
    "end\n",
    "\n",
    "function Base.show(t::Telomere)\n",
    "    return \".\"\n",
    "end\n",
    "\n",
    "# x = Telomere()\n",
    "# show(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b81dd52d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Representation for DNA fragment (a gene) in DCJ model.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    uid: int (required)\n",
    "        unique id\n",
    "\n",
    "    dna: str\n",
    "        What dna fragment does this marker represent?\n",
    "\n",
    "    reverse: boolean\n",
    "        Is the DNA fragment reversed?\n",
    "\"\"\"\n",
    "@with_kw struct Marker <: AbstractMarker\n",
    "    uid::Int\n",
    "    dna::String=\"\"\n",
    "    reverse::Bool=false\n",
    "end\n",
    "\n",
    "function Base.isequal(a::Marker, b::Marker)\n",
    "    return typeof(a) == typeof(b) && a.uid == b.uid\n",
    "end\n",
    "\n",
    "function Base.show(m::Marker)\n",
    "    if m.reverse\n",
    "        print(\"-\" * string(m.uid))\n",
    "    else\n",
    "        print(string(m.uid))\n",
    "    end\n",
    "end\n",
    "\n",
    "# y = Marker(1, \"d\", true)\n",
    "# show(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a0864383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[1mTest Summary: | \u001b[22m\u001b[32m\u001b[1mPass  \u001b[22m\u001b[39m\u001b[36m\u001b[1mTotal  \u001b[22m\u001b[39m\u001b[0m\u001b[1mTime\u001b[22m\n",
      "Markers Tests | \u001b[32m   4  \u001b[39m\u001b[36m    4  \u001b[39m\u001b[0m1.1s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Test.DefaultTestSet(\"Markers Tests\", Any[], 4, false, false, true, 1.717725485875e9, 1.717725486934e9, false, \"In[27]\")"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Test \n",
    "\n",
    "function test_telomere_type()\n",
    "    a = Telomere()\n",
    "    @test typeof(a) == Telomere\n",
    "end\n",
    "\n",
    "function test_forward_reverse_marker()\n",
    "    @test isequal(Marker(0, \"d\", true), Marker(0, \"d\", false))\n",
    "end\n",
    "\n",
    "function test_telomere_repr()\n",
    "    @test show(Telomere()) == \".\"\n",
    "end\n",
    "\n",
    "function test_reverse_equality()\n",
    "    @test !isequal(Marker(0, \"D\", true), Marker(1, \"D\", true))\n",
    "end\n",
    "\n",
    "@testset \"Markers Tests\" begin\n",
    "    test_telomere_type()\n",
    "    test_forward_reverse_marker()\n",
    "    test_telomere_repr()\n",
    "    test_reverse_equality()\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5228f436",
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: `MarkerEnd` not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: `MarkerEnd` not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ In[28]:1"
     ]
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6935f160",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "other_adjacency_end (generic function with 1 method)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Markers with designated head or tail end. Used in adjacencies\n",
    "\"\"\"\n",
    "@with_kw struct MarkerEnd\n",
    "    marker::AbstractMarker\n",
    "    head::Bool=true\n",
    "end\n",
    "\n",
    "\"\"\"Adjacency Data Structure\"\"\"\n",
    "mutable struct Adjacency # TODO markerEnd or marker? i think markerend \n",
    "    left_end_marker::MarkerEnd\n",
    "    right_end_marker::MarkerEnd\n",
    "    label::Vector{Marker}\n",
    "\n",
    "    Adjacency(left_end_marker::MarkerEnd = nothing, right_end_marker::MarkerEnd = nothing; label::Vector{Marker} = nothing) = new([left_end_marker, right_end_marker], left_end_marker, right_end_marker, label)\n",
    "end\n",
    "\n",
    "function Base.isequal(a::MarkerEnd, b::MarkerEnd)\n",
    "    return a.marker == b.marker && a.head == b.head\n",
    "end\n",
    "\n",
    "function Base.show(me::MarkerEnd)\n",
    "    if me.head\n",
    "        print(string(me.marker))\n",
    "    else\n",
    "        print(string(me.marker) * \"*\")\n",
    "    end\n",
    "end\n",
    "\n",
    "function other_adjacency_end(me::MarkerEnd, adj::Adjacency)\n",
    "    if adj.left_end_marker == me\n",
    "        return adj.right_end_marker\n",
    "    else\n",
    "        return adj.left_end_marker\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7f20e9ce",
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "invalid redefinition of constant Main.Chromosome",
     "output_type": "error",
     "traceback": [
      "invalid redefinition of constant Main.Chromosome",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ In[30]:1"
     ]
    }
   ],
   "source": [
    " \"\"\"The abstraction of a chromosome in DCJ model, a list of telomeres and markers\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    markers: list or string (req)\n",
    "        The object to be converted into a chromosome.\n",
    "\n",
    "    \"\"\"\n",
    "mutable struct Chromosome\n",
    "    markers::Vector{AbstractMarker}\n",
    "    marker_set::Set{String}\n",
    "end\n",
    "\n",
    "# mutable struct ChromosomeStatic\n",
    "#     string_d::Dict{String, Int}  # gene to ID \n",
    "#     uid_counter::Int\n",
    "#     ChromosomeStatic() = new(Dict{String, Int}(), 0)\n",
    "# end\n",
    "\n",
    "const chromosome_static = ChromosomeStatic()\n",
    "\n",
    "function Chromosome(markers::Vector{AbstractMarker})\n",
    "    marker_set = Set{AbstractMarker}()\n",
    "    content = AbstractMarker[]\n",
    "    \n",
    "    # If markers is already a list of markers, \n",
    "    if isa(markers, Vector{AbstractMarker})\n",
    "        content = markers\n",
    "     \n",
    "    # throw(TypeError(\"Chromosome must be a list of Marker instances or a string of unique markers.\"))\n",
    "\n",
    "    # Otherwise loop through and parse string\n",
    "    else \n",
    "        markers_str = collect(markers)\n",
    "        if !((markers_str[1] == '.') == (markers_str[end] == '.'))\n",
    "            throw(ArgumentError(\"Linear Chromosome must start and end with telomeres.\"))\n",
    "        end\n",
    "        for (i, s) in enumerate(markers_str)\n",
    "            dna = string(s)\n",
    "            rev = isuppercase(s)\n",
    "            telomere = s == '.'\n",
    "            # handle  marker \n",
    "            if s in marker_set && !telomere\n",
    "                throw(ArgumentError(\"Duplicated markers are not allowed. ($s)\"))\n",
    "            elseif s in keys(chromosome_static.string_d)\n",
    "                uid = chromosome_static.string_d[s]\n",
    "                push!(marker_set, s)\n",
    "            else\n",
    "                chromosome_static.string_d[s] = chromosome_static.uid_counter\n",
    "                push!(marker_set, s)\n",
    "                uid = chromosome_static.uid_counter\n",
    "                chromosome_static.uid_counter += 1\n",
    "            end\n",
    "            # handle telomere\n",
    "            if telomere && !(i == 1 || i == length(markers_str))\n",
    "                throw(ArgumentError(\"Telomere cannot appear in middle of chromosome.\"))\n",
    "            elseif telomere\n",
    "                push!(content, Telomere())\n",
    "            else\n",
    "                push!(content, Marker(uid, dna, rev))\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    return Chromosome(content, marker_set)\n",
    "end\n",
    "\n",
    "function clear_chromosome_static()\n",
    "    chromosome_static.string_d = Dict{String, Int}()\n",
    "    chromosome_static.uid_counter = 0\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88d2634e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Marker(0, \"d\", true), Marker(0, \"d\", false)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "410d12e2",
   "metadata": {},
   "source": [
    "workshopping below....\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99b6e5e3",
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: `AbstractMarker` not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: `AbstractMarker` not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ In[23]:5"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Chromosome\n",
    "\n",
    "mutable struct Chromosome\n",
    "    marker_set::Set{String}\n",
    "    data::Vector{AbstractMarker}\n",
    "    Chromosome(markers::AbstractVector{AbstractMarker}) = new(Set(), markers)\n",
    "    Chromosome(markers::AbstractString) = new(Set(), [])\n",
    "end\n",
    "\n",
    "# Genome\n",
    "\n",
    "mutable struct Genome\n",
    "    data::Vector{Chromosome}\n",
    "    Genome(chromosomes::Chromosome...) = new(chromosomes)\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "# AdjacencyGraph\n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    commonMarkers::Set{AbstractMarker}\n",
    "    adjA::Vector{Adjacency}\n",
    "    adjB::Vector{Adjacency}\n",
    "    cycles::Int\n",
    "    ab_paths::Int\n",
    "    a_runs::Int\n",
    "    b_runs::Int\n",
    "    run_potential::Int\n",
    "    indel_potential::Int\n",
    "end\n",
    "\n",
    "function AdjacencyGraph(A::Genome, B::Genome)\n",
    "    marker_set_a = Set(Iterators.flatten(A.data))\n",
    "    marker_set_b = Set(Iterators.flatten(B.data))\n",
    "    commonMarkers = intersect(marker_set_a, marker_set_b)\n",
    "    if Telomere() in commonMarkers\n",
    "        delete!(commonMarkers, Telomere())\n",
    "    end\n",
    "    marker_set_a = setdiff(marker_set_a, commonMarkers)\n",
    "    marker_set_b = setdiff(marker_set_b, commonMarkers)\n",
    "\n",
    "    adjA = Adjacency[]\n",
    "    adjB = Adjacency[]\n",
    "    adj = [MarkerEnd(Telomere()), MarkerEnd(Telomere())]\n",
    "    adjacencies = [adjA, adjB]\n",
    "    reference_A = Dict{MarkerEnd, Int}(MarkerEnd(Telomere(), true) => 0)\n",
    "    reference_B = Dict{MarkerEnd, Int}(MarkerEnd(Telomere(), true) => 0)\n",
    "    references = [reference_A, reference_B]\n",
    "\n",
    "    for (i, genome) in enumerate([A, B])\n",
    "        adjacency_length = 0\n",
    "        for chromosome in genome.data\n",
    "            index = 0\n",
    "            adj_index = 0\n",
    "            current_marker = chromosome.data[index]\n",
    "            chromosome_markers = Set(chromosome.data)\n",
    "            adjacency_length += length(intersect(commonMarkers, chromosome_markers))\n",
    "            if chromosome.data[1] isa Telomere\n",
    "                adjacency_length += 1\n",
    "            end\n",
    "            while length(adjacencies[i]) < adjacency_length\n",
    "                label = AbstractMarker[]\n",
    "                if current_marker isa Telomere\n",
    "                    adj[1] = MarkerEnd(Telomere(), true)\n",
    "                else\n",
    "                    adj[1] = MarkerEnd(current_marker, !current_marker.reverse)\n",
    "                end\n",
    "                next_marker = ifelse(index >= length(chromosome.data) - 1, chromosome.data[1], chromosome.data[index + 1])\n",
    "                while next_marker ∉ commonMarkers\n",
    "                    if next_marker isa Telomere\n",
    "                        break\n",
    "                    end\n",
    "                    push!(label, next_marker)\n",
    "                    index += 1\n",
    "                    next_marker = ifelse(index >= length(chromosome.data) - 1, chromosome.data[1], chromosome.data[index + 1])\n",
    "                end\n",
    "                if next_marker isa Telomere\n",
    "                    adj[2] = MarkerEnd(Telomere(), true)\n",
    "                else\n",
    "                    adj[2] = MarkerEnd(next_marker, next_marker.reverse)\n",
    "                end\n",
    "                push!(adjacencies[i], Adjacency(adj[1], adj[2], label))\n",
    "                if adj[1].marker != Telomere()\n",
    "                    references[i][adj[1]] = length(adjacencies[i])\n",
    "                end\n",
    "                if adj[2].marker != Telomere()\n",
    "                    references[i][adj[2]] = length(adjacencies[i])\n",
    "                end\n",
    "                current_marker = next_marker\n",
    "                index += 1\n",
    "                adj_index += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    to_visit_a_index = Set(1:length(adjA))\n",
    "    to_visit_b_index = Set(1:length(adjB))\n",
    "    visited_a_index = Set{Int}()\n",
    "    visited_b_index = Set{Int}()\n",
    "    cycles = 0\n",
    "    ab_paths = 0\n",
    "    a_runs = 0\n",
    "    on_a_run = false\n",
    "    b_runs = 0\n",
    "    on_b_run = false\n",
    "\n",
    "    while !isempty(to_visit_a_index)\n",
    "        current_adj_index = pop!(to_visit_a_index)\n",
    "        push!(visited_a_index, current_adj_index)\n",
    "        left_marker = adjA[current_adj_index].left_end_marker\n",
    "        right_marker = adjA[current_adj_index].right_end_marker\n",
    "        if !isempty(adjA[current_adj_index].label)\n",
    "            a_runs += 1\n",
    "            on_a_run = true\n",
    "        end\n",
    "        paths_end_on_a = [true, true]\n",
    "        for (i, current_marker) in enumerate([left_marker, right_marker])\n",
    "            a_side = true\n",
    "            next_adj_index = reference_B[current_marker]\n",
    "            if next_adj_index in visited_b_index\n",
    "                continue\n",
    "            end\n",
    "            while next_adj_index != nothing\n",
    "                current_adj_index = next_adj_index\n",
    "                a_side = !a_side\n",
    "                adj_side = a_side ? adjA : adjB\n",
    "                current_adj = adj_side[current_adj_index]\n",
    "                reference_side = a_side ? reference_B : reference_A\n",
    "                if a_side\n",
    "                    push!(visited_a_index, current_adj_index)\n",
    "                    delete!(to_visit_a_index, current_adj_index)\n",
    "                else\n",
    "                    push!(visited_b_index, current_adj_index)\n",
    "                    delete!(to_visit_b_index, current_adj_index)\n",
    "                end\n",
    "                if !isempty(current_adj.label)\n",
    "                    if !a_side && on_a_run\n",
    "                        on_b_run = true\n",
    "                        on_a_run = false\n",
    "                        b_runs += 1\n",
    "                    elseif !a_side && !on_b_run\n",
    "                        on_b_run = true\n",
    "                        b_runs += 1\n",
    "                    elseif a_side && on_b_run\n",
    "                        on_a_run = true\n",
    "                        on_b_run = false\n",
    "                        a_runs += 1\n",
    "                    elseif a_side && !on_a_run\n",
    "                        on_a_run = true\n",
    "                        a_runs += 1\n",
    "                    end\n",
    "                end\n",
    "                current_marker = current_marker == current_adj.left_end_marker ? current_adj.right_end_marker : current_adj.left_end_marker\n",
    "                next_adj_index = reference_side[current_marker]\n",
    "                if next_adj_index == nothing\n",
    "                    paths_end_on_a[i] = a_side\n",
    "                    break\n",
    "                elseif (a_side && next_adj_index in visited_b_index) || (!a_side && next_adj_index in visited_a_index)\n",
    "                    cycles += 1\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        if paths_end_on_a[1] != paths_end_on_a[2]\n",
    "            ab_paths += 1\n",
    "        end\n",
    "    end\n",
    "\n",
    "    run_potential = a_runs + b_runs\n",
    "    indel_potential = run_potential > 0 ? (run_potential + 1) ÷ 2 + ((run_potential ÷ 2) % 2) : 0\n",
    "\n",
    "    return AdjacencyGraph(commonMarkers, adjA, adjB, cycles, ab_paths, a_runs, b_runs, run_potential, indel_potential)\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "bcd20ebf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "true\n"
     ]
    }
   ],
   "source": [
    "function calculate_distance(A::String, B::String)\n",
    "    check_conditions(A, B)\n",
    "\n",
    "    # Chromosome.clear()\n",
    "    # genome_a = transform_genome(A)\n",
    "    # genome_b = transform_genome(B)\n",
    "    # ag = AdjacencyGraph(genome_a, genome_b)\n",
    "    # dcj_distance = length(ag.commonMarkers) - ag.cycles - ag.ab_paths / 2\n",
    "end\n",
    "\n",
    "# function transform_genome(genome)\n",
    "#     return typeof(genome) == Genome ? genome : Genome([Chromosome(c) for c in genome]...)\n",
    "# end\n",
    "\n",
    "# Chromosome.clear() = (Chromosome.string_d = Dict(), Chromosome.uid_counter = 0)\n",
    "\n",
    "\n",
    "\n",
    "println(calculate_distance(\".abcd.\", \".aCBd.\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b49c8310",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "check_conditions"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Funcs \n",
    "\n",
    "\"\"\"\n",
    "checks for correct conditions \n",
    "    - only letters and dots\n",
    "    - no duplicate genes (letters)\n",
    "    - even number of telomeres in both genomes (can't have the start of a chromosome without closing it)  \n",
    "\"\"\"\n",
    "function check_conditions(A::String, B::String)\n",
    "    set_A = Set{Char}()\n",
    "    set_B = Set{Char}()\n",
    "    tel_count_A = 0 \n",
    "    tel_count_B = 0 \n",
    "\n",
    "    for i in 1:length(A)\n",
    "        if A[i] == '.'\n",
    "            tel_count_A += 1\n",
    "        elseif A[i] ∉ set_A && isletter(A[i])\n",
    "            push!(set_A, lowercase(A[i])) # or uppercase, just keep consistent\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome A has duplicate genes or non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    for i in 1:length(B) \n",
    "        if B[i] == '.'\n",
    "            tel_count_B += 1\n",
    "        elseif  B[i] ∉ set_B  && isletter(A[i])\n",
    "            push!(set_B, lowercase(B[i]))\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome B has duplicate genes or non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    if mod(tel_count_A, 2) != 0 || mod(tel_count_B, 2) != 0  \n",
    "         throw(ArgumentError(\"Error: Telomere error\"))\n",
    "        return false \n",
    "    elseif set_A != set_B\n",
    "        throw(ArgumentError(\"Error: Sets of genes in the two given genomes don't match\"))\n",
    "        return false\n",
    "    end \n",
    "\n",
    "\n",
    "    return true \n",
    "end \n",
    "\n",
    "# correct_conditions(\".abdc.\", \"abCd\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
