{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Funcs \n",
    "\n",
    "\"\"\"\n",
    "checks for correct conditions \n",
    "    - only letters and dots\n",
    "    - even number of telomeres in both genomes (can't have the start of a chromosome without closing it)  \n",
    "\"\"\"\n",
    "function check_input(A::String, B::String)\n",
    "    set_A = Dict{Char, Int64}\n",
    "    set_B = Set{Char}()\n",
    "    tel_count_A = 0 \n",
    "    tel_count_B = 0 \n",
    "\n",
    "    for i in 1:length(A)\n",
    "        if A[i] == '.'\n",
    "            tel_count_A += 1\n",
    "        elseif isletter(A[i])\n",
    "            push!(set_A, lowercase(A[i])) # or uppercase, just keep consistent\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome A has non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    for i in 1:length(B) \n",
    "        if B[i] == '.'\n",
    "            tel_count_B += 1\n",
    "        elseif  isletter(A[i])\n",
    "            push!(set_B, lowercase(B[i]))\n",
    "        else\n",
    "            throw(ArgumentError(\"Error: Genome B has non-letter genes\"))\n",
    "            return false\n",
    "        end \n",
    "    end\n",
    "\n",
    "    if mod(tel_count_A, 2) != 0 || mod(tel_count_B, 2) != 0  \n",
    "         throw(ArgumentError(\"Error: Telomere error\"))\n",
    "        return false \n",
    "    elseif set_A != set_B\n",
    "        throw(ArgumentError(\"Error: Sets of genes in the two given genomes don't match\"))\n",
    "        return false\n",
    "    end \n",
    "\n",
    "\n",
    "    return true \n",
    "end \n",
    "\n",
    "# correct_conditions(\".abdc.\", \"abCd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function Chromosome(genes::Vector{AbstractGene}, id_set::Set{Int})\n",
    "    for gene in genes\n",
    "        telomere = show(gene) == \".\"\n",
    "        if !telomere \n",
    "            if gene.id âˆ‰ id_set \n",
    "                push!(id_set, gene.id)\n",
    "            else \n",
    "                throw(ArgumentError(\"There are duplicate genes with the same ID in this genome.\"))\n",
    "            end\n",
    "        end \n",
    "    end \n",
    "    content = genes\n",
    "\n",
    "    return Chromosome(content)\n",
    "\n",
    "end \n",
    "\n",
    "\n",
    "## inside Chromosome\n",
    "# check if empty chromosome .. \n",
    "if g == '.' \n",
    "    chrom = chrom * g \n",
    "    push!(chrom_list, chrom)\n",
    "    chrom = \"\"\n",
    "    continue \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## tests telomeres and genes \n",
    "\n",
    "using Test \n",
    "\n",
    "function test_telomere_type()\n",
    "    a = Telomere()\n",
    "    @test typeof(a) == Telomere\n",
    "end\n",
    "\n",
    "function test_forward_reverse_gene()\n",
    "    @test isequal(Gene(0, \"d\", true), Gene(0, \"d\", false))\n",
    "end\n",
    "\n",
    "function test_telomere_repr()\n",
    "    @test show(Telomere()) == \".\"\n",
    "end\n",
    "\n",
    "function test_reverse_equality()\n",
    "    @test !isequal(Gene(0, \"D\", true), Gene(1, \"D\", true))\n",
    "end\n",
    "\n",
    "@testset \"Gene Tests\" begin\n",
    "    test_telomere_type()\n",
    "    test_forward_reverse_gene()\n",
    "    test_telomere_repr()\n",
    "    test_reverse_equality()\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Test\n",
    "\n",
    "function test_telomere_type()\n",
    "    a = Telomere()\n",
    "    @test typeof(a) == Telomere\n",
    "end\n",
    "\n",
    "function test_forward_reverse_marker()\n",
    "    @test Marker(0, \"d\", true) == Marker(0, \"d\", false)\n",
    "end\n",
    "\n",
    "function test_telomere_repr()\n",
    "    @test string(Telomere()) == \".\"\n",
    "end\n",
    "\n",
    "function test_reverse_equality()\n",
    "    @test Marker(0, \"D\", true) != Marker(1, \"D\", true)\n",
    "end\n",
    "\n",
    "function test_dcj_single_a_dcj()\n",
    "    a = Genome(Chromosome(\".ac.\"))\n",
    "    b = Genome(Chromosome(\".abc.\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 0\n",
    "end\n",
    "\n",
    "function test_dcj_single_swap_dcj()\n",
    "    a = Genome(Chromosome(\".acb.\"))\n",
    "    b = Genome(Chromosome(\".abc.\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 2\n",
    "end\n",
    "\n",
    "function test_dcj_intermediate_dcj()\n",
    "    a = Genome(Chromosome(\".cbAdEGf.\"))\n",
    "    b = Genome(Chromosome(\".abcdefg.\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 6\n",
    "end\n",
    "\n",
    "function test_dcj_intermediate_2_dcj()\n",
    "    a = Genome(Chromosome(\".aBdCE.\"))\n",
    "    b = Genome(Chromosome(\".abcde.\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 4\n",
    "end\n",
    "\n",
    "function test_dcj_backwards()\n",
    "    a = Genome(Chromosome(\"abc\"))\n",
    "    b = Genome(Chromosome(\"cba\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 2\n",
    "end\n",
    "\n",
    "function test_dcj_single_split()\n",
    "    a = Genome(Chromosome(\".ab.\"))\n",
    "    b = Genome(Chromosome(\"ab\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 1\n",
    "end\n",
    "\n",
    "function test_dcj_circular_excision()\n",
    "    a = Genome(Chromosome(\"ab\"), Chromosome(\".cd.\"))\n",
    "    b = Genome(Chromosome(\".abcd.\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 1\n",
    "end\n",
    "\n",
    "function test_dcj_default_example()\n",
    "    a = Genome(Chromosome(\"ab\"), Chromosome(\".cd.\"), Chromosome(\".e.\"), Chromosome(\"fg\"))\n",
    "    b = Genome(Chromosome(\".acD.\"), Chromosome(\"be\"), Chromosome(\".fg.\"))\n",
    "    @test calculate_distance(a, b, method=\"dcj\") == 5\n",
    "end\n",
    "\n",
    "function test_duplicated_markers()\n",
    "    @test_throws ValueError transform_genome([\"aa\"])\n",
    "end\n",
    "\n",
    "function test_duplicated_marker_reversed()\n",
    "    @test_throws ValueError transform_genome([\"Aa\"])\n",
    "end\n",
    "\n",
    "function test_duplicated_marker_between_chromosomes()\n",
    "    @test_throws ValueError transform_genome([\"abcde\", \".fghiC.\"])\n",
    "end\n",
    "\n",
    "function test_missing_telomere()\n",
    "    @test_throws ValueError transform_genome([\".abcd\"])\n",
    "    @test_throws ValueError transform_genome([\"abcd.\"])\n",
    "end\n",
    "\n",
    "function test_rogue_telomere()\n",
    "    @test_throws ValueError transform_genome([\".abc.de.\"])\n",
    "end\n",
    "\n",
    "function test_genome_without_chromosomes()\n",
    "    @test_throws TypeError transform_genome(\"abcde\", \".fg.\")\n",
    "end\n",
    "\n",
    "function test_chromosome_list_not_markers()\n",
    "    @test_throws TypeError transform_genome([[\"a\", \"b\", \"c\"]])\n",
    "end\n",
    "\n",
    "@testset \"Markers Tests\" begin\n",
    "    test_telomere_type()\n",
    "    test_forward_reverse_marker()\n",
    "    test_telomere_repr()\n",
    "    test_reverse_equality()\n",
    "end\n",
    "\n",
    "@testset \"Adjacency Graph Linear Tests\" begin\n",
    "    test_dcj_single_a_dcj()\n",
    "    test_dcj_single_swap_dcj()\n",
    "    test_dcj_intermediate_dcj()\n",
    "    test_dcj_intermediate_2_dcj()\n",
    "end\n",
    "\n",
    "@testset \"Adjacency Graph Circular Tests\" begin\n",
    "    test_dcj_backwards()\n",
    "    test_dcj_single_split()\n",
    "    test_dcj_circular_excision()\n",
    "    test_dcj_default_example()\n",
    "end\n",
    "\n",
    "@testset \"Genome Validation Tests\" begin\n",
    "    test_duplicated_markers()\n",
    "    test_duplicated_marker_reversed()\n",
    "    test_duplicated_marker_between_chromosomes()\n",
    "    test_missing_telomere()\n",
    "    test_rogue_telomere()\n",
    "    test_genome_without_chromosomes()\n",
    "    test_chromosome_list_not_markers()\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function reset_documentation(id_counter::Ref{Int}, id_to_char::Dict{Int, Char}, char_to_id::Dict{Char, Int})\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}() \n",
    "end "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Test\n",
    "\n",
    "\"\"\" TESTING \"\"\" \n",
    "### note: circular chroms start and end with the same letter (case-sensitive)\n",
    "\n",
    "\n",
    "function test_single_gene_chrom_linear()\n",
    "    # str_to_genome func params\n",
    "    id_counter = Ref{Int}(1)\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}() \n",
    "\n",
    "    # genome setup\n",
    "    data = Vector{Chromosome}()\n",
    "    id_to_char = Dict{Int, Char}()\n",
    "    char_to_id = Dict{Char, Int}()\n",
    "\n",
    "    id_to_char[1] = 'a'\n",
    "    char_to_id['a'] = 1\n",
    "    chrom = Chromosome([Telomere(), Gene(1, 'a', false), Telomere()])\n",
    "    data = [chrom]\n",
    "\n",
    "    @test string_to_genome(\"a\", id_counter, id_to_char, char_to_id, true) == Genome(data, id_to_char, char_to_id)\n",
    "end\n",
    "\n",
    "\n",
    "@testset \"String to Genome Tests\" begin\n",
    "    test_single_gene_chrom_linear()\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using NBInclude\n",
    "@nbinclude(\"dcj_algo.ipynb\")\n",
    "\n",
    "# id_counter = Ref{Int}(1)\n",
    "# id_to_char = Dict{Int, Char}()\n",
    "# char_to_id = Dict{Char, Int}()\n",
    "\n",
    "# s = \".a.\"\n",
    "\n",
    "# genome = string_to_genome(s, id_counter, id_to_char, char_to_id, true)\n",
    "\n",
    "# # show(genome)\n",
    "\n",
    "# adj_list, adj_set = genome_to_adj_listset(genome)   \n",
    "\n",
    "# show(adj_list)\n",
    "\n",
    "# 2 DCJ op - adj --> adj  \n",
    "src = \".abc.\"\n",
    "target = \".acb.\"\n",
    "\n",
    "# 1 DCJ op - adj --> telo  \n",
    "src = \"abca\"\n",
    "target = \".abc.\"\n",
    "\n",
    "# 1 DCJ op - telo --> adj  \n",
    "src = \".ab.cc\"\n",
    "target = \".abc.\"\n",
    "\n",
    "\n",
    "\n",
    "src = \".abcdefg.\"\n",
    "target = \".abcdGFE.\" \n",
    "\n",
    "calculate_distance(src, target)\n",
    "\n",
    "\n",
    "\n",
    "# single chrom, single gene \n",
    "s = \".a.\"\n",
    "s = \"aa\" \n",
    "\n",
    "# single chrom, 2 genes\n",
    "s = \".ab.\"\n",
    "s = \"aba\"\n",
    "\n",
    "# single chrom, 3 genes\n",
    "s = \".abc.\"\n",
    "s = \"abca\"\n",
    "\n",
    "# multiple chrom, multiple genes \n",
    "s = \".a.bb.c.\"\n",
    "s = \"aa.b.cc\"\n",
    "s = \".ab.cdc.de.fgf\"\n",
    "s = \"aba.cd.ded.fg.\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
