{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  AdjacencyGraph\n",
    "\n",
    "mutable struct AdjacencyGraph\n",
    "    commonGenes::Set{AbstractGene}\n",
    "    adjA::Vector{Adjacency}\n",
    "    adjB::Vector{Adjacency}\n",
    "    cycles::Int\n",
    "    ab_paths::Int\n",
    "    a_runs::Int\n",
    "    b_runs::Int\n",
    "    run_potential::Int\n",
    "    indel_potential::Int\n",
    "end\n",
    "\n",
    "function AdjacencyGraph(A::Genome, B::Genome)\n",
    "    gene_set_a = Set(Iterators.flatten(A.data))\n",
    "    gene_set_b = Set(Iterators.flatten(B.data))\n",
    "    commonGenes = intersect(gene_set_a, gene_set_b)\n",
    "    if Telomere() in commonGenes\n",
    "        delete!(commonGenes, Telomere())\n",
    "    end\n",
    "    gene_set_a = setdiff(gene_set_a, commonGenes)\n",
    "    gene_set_b = setdiff(gene_set_b, commonGenes)\n",
    "\n",
    "    adjA = Adjacency[]\n",
    "    adjB = Adjacency[]\n",
    "    adj = [GeneEnd(Telomere()), GeneEnd(Telomere())]\n",
    "    adjacencies = [adjA, adjB]\n",
    "    reference_A = Dict{GeneEnd, Int}(GeneEnd(Telomere(), true) => 0)\n",
    "    reference_B = Dict{GeneEnd, Int}(GeneEnd(Telomere(), true) => 0)\n",
    "    references = [reference_A, reference_B]\n",
    "\n",
    "    for (i, genome) in enumerate([A, B])\n",
    "        adjacency_length = 0\n",
    "        for chromosome in genome.data\n",
    "            index = 0\n",
    "            adj_index = 0\n",
    "            current_gene = chromosome.data[index]\n",
    "            chromosome_genes = Set(chromosome.data)\n",
    "            adjacency_length += length(intersect(commonGenes, chromosome_Genes))\n",
    "            if chromosome.data[1] isa Telomere\n",
    "                adjacency_length += 1\n",
    "            end\n",
    "            while length(adjacencies[i]) < adjacency_length\n",
    "                label = AbstractGene[]\n",
    "                if current_gene isa Telomere\n",
    "                    adj[1] = GeneEnd(Telomere(), true)\n",
    "                else\n",
    "                    adj[1] = GeneEnd(current_gene, !current_gene.reverse)\n",
    "                end\n",
    "                next_gene = ifelse(index >= length(chromosome.data) - 1, chromosome.data[1], chromosome.data[index + 1])\n",
    "                while next_gene ∉ commonGene\n",
    "                    if next_gene isa Telomere\n",
    "                        break\n",
    "                    end\n",
    "                    push!(label, next_gene)\n",
    "                    index += 1\n",
    "                    next_gene = ifelse(index >= length(chromosome.data) - 1, chromosome.data[1], chromosome.data[index + 1])\n",
    "                end\n",
    "                if next_gene isa Telomere\n",
    "                    adj[2] = GeneEnd(Telomere(), true)\n",
    "                else\n",
    "                    adj[2] = GeneEnd(next_gene, next_gene.reverse)\n",
    "                end\n",
    "                push!(adjacencies[i], Adjacency(adj[1], adj[2], label))\n",
    "                if adj[1].gene != Telomere()\n",
    "                    references[i][adj[1]] = length(adjacencies[i])\n",
    "                end\n",
    "                if adj[2].gene != Telomere()\n",
    "                    references[i][adj[2]] = length(adjacencies[i])\n",
    "                end\n",
    "                current_gene = next_gene\n",
    "                index += 1\n",
    "                adj_index += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # traversing adj graph \n",
    "    to_visit_a_index = Set(1:length(adjA))\n",
    "    to_visit_b_index = Set(1:length(adjB))\n",
    "    visited_a_index = Set{Int}()\n",
    "    visited_b_index = Set{Int}()\n",
    "    cycles = 0\n",
    "    ab_paths = 0\n",
    "    a_runs = 0\n",
    "    on_a_run = false\n",
    "    b_runs = 0\n",
    "    on_b_run = false\n",
    "\n",
    "    while !isempty(to_visit_a_index)\n",
    "        current_adj_index = pop!(to_visit_a_index)\n",
    "        push!(visited_a_index, current_adj_index)\n",
    "        left_gene = adjA[current_adj_index].left_end_gene\n",
    "        right_gene = adjA[current_adj_index].right_end_gene\n",
    "        if !isempty(adjA[current_adj_index].label)\n",
    "            a_runs += 1\n",
    "            on_a_run = true\n",
    "        end\n",
    "        paths_end_on_a = [true, true]\n",
    "        for (i, current_gene) in enumerate([left_gene, right_gene])\n",
    "            a_side = true\n",
    "            next_adj_index = reference_B[current_gene]\n",
    "            if next_adj_index in visited_b_index\n",
    "                continue\n",
    "            end\n",
    "            while next_adj_index != nothing\n",
    "                current_adj_index = next_adj_index\n",
    "                a_side = !a_side\n",
    "                adj_side = a_side ? adjA : adjB\n",
    "                current_adj = adj_side[current_adj_index]\n",
    "                reference_side = a_side ? reference_B : reference_A\n",
    "                if a_side\n",
    "                    push!(visited_a_index, current_adj_index)\n",
    "                    delete!(to_visit_a_index, current_adj_index)\n",
    "                else\n",
    "                    push!(visited_b_index, current_adj_index)\n",
    "                    delete!(to_visit_b_index, current_adj_index)\n",
    "                end\n",
    "                if !isempty(current_adj.label)\n",
    "                    if !a_side && on_a_run\n",
    "                        on_b_run = true\n",
    "                        on_a_run = false\n",
    "                        b_runs += 1\n",
    "                    elseif !a_side && !on_b_run\n",
    "                        on_b_run = true\n",
    "                        b_runs += 1\n",
    "                    elseif a_side && on_b_run\n",
    "                        on_a_run = true\n",
    "                        on_b_run = false\n",
    "                        a_runs += 1\n",
    "                    elseif a_side && !on_a_run\n",
    "                        on_a_run = true\n",
    "                        a_runs += 1\n",
    "                    end\n",
    "                end\n",
    "                current_gene = current_gene == current_adj.left_end_gene ? current_adj.right_end_gene : current_adj.left_end_gene\n",
    "                next_adj_index = reference_side[current_gene]\n",
    "                if next_adj_index == nothing\n",
    "                    paths_end_on_a[i] = a_side\n",
    "                    break\n",
    "                elseif (a_side && next_adj_index in visited_b_index) || (!a_side && next_adj_index in visited_a_index)\n",
    "                    cycles += 1\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        if paths_end_on_a[1] != paths_end_on_a[2]\n",
    "            ab_paths += 1\n",
    "        end\n",
    "    end\n",
    "\n",
    "    run_potential = a_runs + b_runs\n",
    "    indel_potential = run_potential > 0 ? (run_potential + 1) ÷ 2 + ((run_potential ÷ 2) % 2) : 0\n",
    "\n",
    "    return AdjacencyGraph(commonGenes, adjA, adjB, cycles, ab_paths, a_runs, b_runs, run_potential, indel_potential)\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pseudo for algo 1 - construction of adj graph \n",
    "\n",
    "# src_location_to_adj = Dict{Int, Vector{GeneEnd}}() \n",
    "# src_gene_to_location = Dict{Gene, Vector{Int}}()\n",
    "\n",
    "# target_location_to_adj = Dict{Int, Vector{GeneEnd}}() \n",
    "# target_gene_to_location = Dict{Gene, Vector{Int}}()\n",
    "\n",
    "# for each adj {p, q} in src genome \n",
    "    # create edge connecting m, {p, q} and vertex of genome B that contains p \n",
    "    # create edge connecting {p, q} and vertex of genome B that contains q\n",
    "\n",
    "\n",
    "# for each telomere {p} in src genome \n",
    "    # create edge connecting {p} and the vertex of target that contains p\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pseudo for algo 2 - dcj operations and distance\n",
    "\n",
    "# retain count \n",
    "\n",
    "# for each adj {p, q} in genome B \n",
    "#     let u be element of genome A that contains p \n",
    "#     let v be element of genome A that contains q \n",
    "#     if u != v \n",
    "#         replace u and v in A by {p, q} and (u\\{p}) U (v\\{q})\n",
    "#         count++ \n",
    "\n",
    "# for each telomere {p} in genome B \n",
    "#     let u be element of genome A that contains p \n",
    "#     if u is an adjacency then \n",
    "#         replace u in A by {p} and (u\\{p})\n",
    "#         count++ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "  # converting char to idx in alphabet \n",
    "    # Int.(['a','c']) .- Int('a') .+ 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"function combine_ge(adj_list::Vector{Adjacency}, u::GeneEnd, u_idx::Int, v::GeneEnd, v_idx::Int, otherge::Bool)\n",
    "    if u_idx < v_idx \n",
    "        u_lt_v = true \n",
    "    else \n",
    "        u_lt_v = false \n",
    "    end \n",
    "\n",
    "    u_adj = src_adj_list[u_idx]\n",
    "    v_adj = src_adj_list[v_idx]\n",
    "    \n",
    "    \n",
    "    if otherge  # (u\\{p}) U (v\\{q}) \n",
    "        new_u = other_adjacency_end(u, u_adj)\n",
    "        new_v = other_adjacency_end(v, v_adj)\n",
    "    else  # {u, v} retaining original order \n",
    "\n",
    "    end \n",
    "\n",
    "    if !u_lt_v \n",
    "        return Adjacency(new_u, new_v)\n",
    "    else \n",
    "        return Adjacency(new_v, new_u)\n",
    "    end\n",
    "\n",
    "end \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "BoundsError: attempt to access 3-element Vector{Int64} at index [0]",
     "output_type": "error",
     "traceback": [
      "BoundsError: attempt to access 3-element Vector{Int64} at index [0]",
      "",
      "Stacktrace:",
      " [1] getindex",
      "   @ .\\essentials.jl:13 [inlined]",
      " [2] nth_permutation(n::Int64, elements::Vector{Int64})",
      "   @ Main .\\In[20]:27",
      " [3] top-level scope",
      "   @ In[20]:36"
     ]
    }
   ],
   "source": [
    "function next_permutation(perm)\n",
    "  \"\"\"\n",
    "  Finds the next permutation of a list in lexicographical order.\n",
    "\n",
    "  Args:\n",
    "      perm: A list of integers representing a permutation.\n",
    "\n",
    "  Returns:\n",
    "      The next permutation of the input list, or None if there is no next permutation.\n",
    "  \"\"\"\n",
    "  n = length(perm)\n",
    "  i = n - 2\n",
    "  while i >= 1 && perm[i] >= perm[i + 1]\n",
    "    i -= 1\n",
    "  end\n",
    "  if i == 0\n",
    "    return nothing\n",
    "  end\n",
    "  j = n - 1\n",
    "  while j > i && perm[j] <= perm[i]\n",
    "    j -= 1\n",
    "  end\n",
    "  perm[i], perm[j] = perm[j], perm[i]\n",
    "  reverse!(perm[i + 1:end])\n",
    "  return perm\n",
    "end\n",
    "\n",
    "function nth_permutation(n, elements)\n",
    "  \"\"\"\n",
    "  Generates the nth permutation of a list of elements in lexicographical order.\n",
    "\n",
    "  Args:\n",
    "      n: The index of the desired permutation (1-based).\n",
    "      elements: A list of unique integers representing the elements to permute.\n",
    "\n",
    "  Returns:\n",
    "      The nth permutation of the input elements.\n",
    "  \"\"\"\n",
    "  perm = elements\n",
    "  n -= 1\n",
    "  while n > 0\n",
    "    perm = next_permutation(perm)\n",
    "    if perm == nothing\n",
    "      break\n",
    "    end\n",
    "    n -= 1\n",
    "  end\n",
    "  return perm\n",
    "end\n",
    "\n",
    "# Example usage\n",
    "elements = [1, 2, 3]\n",
    "first_permutation = nth_permutation(1, elements)\n",
    "second_permutation = nth_permutation(2, elements)\n",
    "\n",
    "println(first_permutation)  # Output: [1, 2, 3]\n",
    "println(second_permutation) # Output: [1, 3, 2]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
